// @agent-relay/sdk/dist/workflows/collectors/codex.js
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);
const CODEX_HOME = path.join(os.homedir(), '.codex');
const DEFAULT_HISTORY_PATH = path.join(CODEX_HOME, 'history.jsonl');
const DEFAULT_STATE_PATH = path.join(CODEX_HOME, 'state_5.sqlite');
function loadBetterSqlite3() {
    try {
        return require('better-sqlite3');
    }
    catch {
        return null;
    }
}
async function openDatabase(dbPath) {
    const BetterSqlite = loadBetterSqlite3();
    if (BetterSqlite) {
        try {
            return new BetterSqlite(dbPath, { readonly: true, fileMustExist: true });
        }
        catch {
            // Fall through to node:sqlite.
        }
    }
    try {
        const sqlite = (await import('node:sqlite'));
        return new sqlite.DatabaseSync(dbPath, { readOnly: true, open: true });
    }
    catch {
        return null;
    }
}
function normalizeTimestamp(value) {
    const numeric = typeof value === 'number' && Number.isFinite(value)
        ? value
        : typeof value === 'string' && value.trim()
            ? Number(value)
            : null;
    if (numeric === null || !Number.isFinite(numeric)) {
        return null;
    }
    return numeric < 10_000_000_000 ? numeric * 1000 : numeric;
}
function parseJsonLine(line) {
    try {
        return JSON.parse(line);
    }
    catch {
        return null;
    }
}
function parseModelProvider(value) {
    if (!value) {
        return { provider: null, model: null };
    }
    if (value.includes('/')) {
        const [provider, ...rest] = value.split('/');
        return {
            provider: provider || null,
            model: rest.join('/') || null,
        };
    }
    if (value.includes(':')) {
        const [provider, ...rest] = value.split(':');
        return {
            provider: provider || null,
            model: rest.join(':') || null,
        };
    }
    return {
        provider: value,
        model: null,
    };
}
function getNumericField(row, fieldNames) {
    for (const fieldName of fieldNames) {
        const value = row[fieldName];
        if (typeof value === 'number' && Number.isFinite(value)) {
            return value;
        }
    }
    return null;
}
function extractTokens(row) {
    const input = getNumericField(row, ['input_tokens', 'prompt_tokens', 'tokens_input']);
    const output = getNumericField(row, ['output_tokens', 'completion_tokens', 'tokens_output']);
    const cacheRead = getNumericField(row, ['cache_read_tokens', 'tokens_cache_read', 'cached_input_tokens']);
    if (input !== null || output !== null || cacheRead !== null) {
        return {
            input: input ?? 0,
            output: output ?? 0,
            cacheRead: cacheRead ?? 0,
        };
    }
    return typeof row.tokens_used === 'number'
        ? {
            input: row.tokens_used,
            output: 0,
            cacheRead: 0,
        }
        : null;
}
export class CodexCollector {
    historyPath;
    statePath;
    constructor(options = {}) {
        this.historyPath = options.historyPath ?? DEFAULT_HISTORY_PATH;
        this.statePath = options.statePath ?? DEFAULT_STATE_PATH;
    }
    canCollect() {
        return fs.existsSync(this.statePath) || fs.existsSync(this.historyPath);
    }
    async collect(query) {
        const historyEntries = this.readHistoryEntries();
        const matchedThread = await this.findMatchingThread(query);
        if (matchedThread) {
            const errors = await this.readThreadErrors(matchedThread.id);
            const { provider, model } = parseModelProvider(matchedThread.model_provider);
            const createdAtMs = normalizeTimestamp(matchedThread.created_at);
            const updatedAtMs = normalizeTimestamp(matchedThread.updated_at);
            return {
                cli: 'codex',
                sessionId: matchedThread.id,
                model,
                provider,
                durationMs: createdAtMs !== null && updatedAtMs !== null && updatedAtMs >= createdAtMs
                    ? updatedAtMs - createdAtMs
                    : Math.max(query.completedAt - query.startedAt, 0),
                cost: null,
                tokens: extractTokens(matchedThread),
                turns: historyEntries.filter((entry) => entry.session_id === matchedThread.id).length,
                toolCalls: [],
                errors,
                finalStatus: errors.length > 0 ? 'failed' : 'unknown',
                summary: null,
                raw: {
                    matchedVia: 'threads',
                    thread: matchedThread,
                },
            };
        }
        const historyMatch = this.findMatchingHistoryEntry(query, historyEntries);
        if (!historyMatch) {
            return null;
        }
        return {
            cli: 'codex',
            sessionId: historyMatch.session_id ?? null,
            model: null,
            provider: null,
            durationMs: Math.max(query.completedAt - query.startedAt, 0),
            cost: null,
            tokens: null,
            turns: historyMatch.session_id
                ? historyEntries.filter((entry) => entry.session_id === historyMatch.session_id).length
                : 0,
            toolCalls: [],
            errors: [],
            finalStatus: 'unknown',
            summary: null,
            raw: {
                matchedVia: 'history',
                entry: historyMatch,
            },
        };
    }
    readHistoryEntries() {
        if (!fs.existsSync(this.historyPath)) {
            return [];
        }
        try {
            return fs.readFileSync(this.historyPath, 'utf8')
                .split(/\r?\n/)
                .map((line) => line.trim())
                .filter(Boolean)
                .flatMap((line) => {
                const parsed = parseJsonLine(line);
                return parsed ? [parsed] : [];
            });
        }
        catch {
            return [];
        }
    }
    findMatchingHistoryEntry(query, entries) {
        for (let index = entries.length - 1; index >= 0; index -= 1) {
            const entry = entries[index];
            const timestamp = normalizeTimestamp(entry.ts);
            if (timestamp === null) {
                continue;
            }
            if (timestamp >= query.startedAt && timestamp <= query.completedAt) {
                return entry;
            }
        }
        return null;
    }
    async findMatchingThread(query) {
        if (!fs.existsSync(this.statePath)) {
            return null;
        }
        const db = await openDatabase(this.statePath);
        if (!db) {
            return null;
        }
        try {
            const threads = db.prepare(`
          SELECT *
          FROM threads
          WHERE cwd = ?
          ORDER BY created_at DESC
          LIMIT 100
        `).all(query.cwd);
            return threads.find((thread) => {
                const createdAt = normalizeTimestamp(thread.created_at);
                return createdAt !== null && createdAt >= query.startedAt && createdAt <= query.completedAt;
            }) ?? null;
        }
        catch {
            return null;
        }
        finally {
            db.close?.();
        }
    }
    async readThreadErrors(threadId) {
        if (!fs.existsSync(this.statePath)) {
            return [];
        }
        const db = await openDatabase(this.statePath);
        if (!db) {
            return [];
        }
        try {
            const rows = db.prepare(`
          SELECT ts, level, message, line
          FROM logs
          WHERE thread_id = ?
            AND lower(level) = 'error'
          ORDER BY ts ASC
        `).all(threadId);
            return rows
                .map((row, index) => {
                const message = typeof row.message === 'string' ? row.message.trim() : '';
                if (!message) {
                    return null;
                }
                return {
                    turn: index + 1,
                    text: message,
                };
            })
                .filter((row) => row !== null);
        }
        catch {
            return [];
        }
        finally {
            db.close?.();
        }
    }
}
