// @relaycast/types/dist/command.js
import { z } from 'zod';
export const CommandParameterSchema = z.object({
    name: z.string(),
    description: z.string().optional(),
    type: z.enum(['string', 'number', 'boolean']),
    required: z.boolean().optional(),
});
export const AgentCommandSchema = z.object({
    id: z.string(),
    command: z.string(),
    description: z.string(),
    handler_agent: z.string(),
    parameters: z.array(CommandParameterSchema),
    created_at: z.string(),
    is_active: z.boolean(),
});
export const CreateCommandRequestSchema = z.object({
    command: z.string(),
    description: z.string(),
    handler_agent: z.string(),
    parameters: z.array(CommandParameterSchema).optional(),
});
export const CreateCommandResponseSchema = z.object({
    id: z.string(),
    command: z.string(),
    description: z.string(),
    handler_agent: z.string(),
    parameters: z.array(CommandParameterSchema),
    created_at: z.string(),
});
export const InvokeCommandRequestSchema = z.object({
    channel: z.string(),
    args: z.string().optional(),
    parameters: z.record(z.string(), z.unknown()).optional(),
});
export const CommandInvocationSchema = z.object({
    id: z.string(),
    command: z.string(),
    channel: z.string(),
    invoked_by: z.string(),
    handler_agent_id: z.string(),
    args: z.string().nullable(),
    parameters: z.record(z.string(), z.unknown()).nullable(),
    message_id: z.string(),
    created_at: z.string(),
});
