// @relaycast/types/dist/events.js
import { z } from 'zod';
import { CoreMessagePayloadSchema } from './message.js';
import { FileAttachmentSchema } from './file.js';
// WebSocket client -> server
export const SubscribeEventSchema = z.object({
    type: z.literal('subscribe'),
    channels: z.array(z.string()),
});
export const UnsubscribeEventSchema = z.object({
    type: z.literal('unsubscribe'),
    channels: z.array(z.string()),
});
export const PingEventSchema = z.object({
    type: z.literal('ping'),
});
export const ClientEventSchema = z.discriminatedUnion('type', [
    SubscribeEventSchema,
    UnsubscribeEventSchema,
    PingEventSchema,
]);
// WebSocket server -> client
export const ChannelMessagePayloadSchema = CoreMessagePayloadSchema.extend({
    attachments: z.array(FileAttachmentSchema),
});
export const MessageCreatedEventSchema = z.object({
    type: z.literal('message.created'),
    channel: z.string(),
    message: ChannelMessagePayloadSchema,
});
export const MessageUpdatedEventSchema = z.object({
    type: z.literal('message.updated'),
    channel: z.string(),
    message: CoreMessagePayloadSchema,
});
export const ThreadReplyEventSchema = z.object({
    type: z.literal('thread.reply'),
    channel: z.string(),
    parent_id: z.string(),
    message: CoreMessagePayloadSchema,
});
export const ReactionAddedEventSchema = z.object({
    type: z.literal('reaction.added'),
    message_id: z.string(),
    emoji: z.string(),
    agent_name: z.string(),
});
export const ReactionRemovedEventSchema = z.object({
    type: z.literal('reaction.removed'),
    message_id: z.string(),
    emoji: z.string(),
    agent_name: z.string(),
});
export const DmReceivedEventSchema = z.object({
    type: z.literal('dm.received'),
    conversation_id: z.string(),
    message: CoreMessagePayloadSchema,
});
export const GroupDmReceivedEventSchema = z.object({
    type: z.literal('group_dm.received'),
    conversation_id: z.string(),
    message: CoreMessagePayloadSchema,
});
export const AgentOnlineEventSchema = z.object({
    type: z.literal('agent.online'),
    agent: z.object({ name: z.string() }),
});
export const AgentOfflineEventSchema = z.object({
    type: z.literal('agent.offline'),
    agent: z.object({ name: z.string() }),
});
export const AgentSpawnRequestedEventSchema = z.object({
    type: z.literal('agent.spawn_requested'),
    agent: z.object({
        name: z.string(),
        cli: z.string(),
        task: z.string(),
        channel: z.string().nullable(),
        already_existed: z.boolean(),
    }),
});
export const AgentReleaseRequestedEventSchema = z.object({
    type: z.literal('agent.release_requested'),
    agent: z.object({ name: z.string() }),
    reason: z.string().nullable(),
    deleted: z.boolean(),
});
export const ChannelCreatedEventSchema = z.object({
    type: z.literal('channel.created'),
    channel: z.object({ name: z.string(), topic: z.string().nullable() }),
});
export const ChannelUpdatedEventSchema = z.object({
    type: z.literal('channel.updated'),
    channel: z.object({ name: z.string(), topic: z.string().nullable() }),
});
export const ChannelArchivedEventSchema = z.object({
    type: z.literal('channel.archived'),
    channel: z.object({ name: z.string() }),
});
export const MemberJoinedEventSchema = z.object({
    type: z.literal('member.joined'),
    channel: z.string(),
    agent_name: z.string(),
});
export const MemberLeftEventSchema = z.object({
    type: z.literal('member.left'),
    channel: z.string(),
    agent_name: z.string(),
});
export const ChannelMutedEventSchema = z.object({
    type: z.literal('member.channel_muted'),
    channel: z.string(),
    agent_name: z.string(),
});
export const ChannelUnmutedEventSchema = z.object({
    type: z.literal('member.channel_unmuted'),
    channel: z.string(),
    agent_name: z.string(),
});
export const MessageReadEventSchema = z.object({
    type: z.literal('message.read'),
    message_id: z.string(),
    agent_name: z.string(),
    read_at: z.string(),
});
export const FileUploadedEventSchema = z.object({
    type: z.literal('file.uploaded'),
    file: z.object({ file_id: z.string(), filename: z.string(), uploaded_by: z.string() }),
});
export const PongEventSchema = z.object({
    type: z.literal('pong'),
});
export const WebhookReceivedEventSchema = z.object({
    type: z.literal('webhook.received'),
    webhook_id: z.string(),
    channel: z.string(),
    message: z.object({ id: z.string(), text: z.string(), source: z.string().nullable() }),
});
export const CommandInvokedEventSchema = z.object({
    type: z.literal('command.invoked'),
    command: z.string(),
    channel: z.string(),
    invoked_by: z.string(),
    handler_agent_id: z.string(),
    args: z.string().nullable(),
    parameters: z.record(z.string(), z.unknown()).nullable(),
});
// WebSocket client events (emitted by WsClient, not from server)
export const WsOpenEventSchema = z.object({
    type: z.literal('open'),
});
export const WsErrorEventSchema = z.object({
    type: z.literal('error'),
});
export const WsReconnectingEventSchema = z.object({
    type: z.literal('reconnecting'),
    attempt: z.number(),
});
export const WsPermanentlyDisconnectedEventSchema = z.object({
    type: z.literal('permanently_disconnected'),
    attempt: z.number(),
});
export const WsCloseEventSchema = z.object({
    type: z.literal('close'),
});
export const ServerEventSchema = z.discriminatedUnion('type', [
    MessageCreatedEventSchema,
    MessageUpdatedEventSchema,
    ThreadReplyEventSchema,
    ReactionAddedEventSchema,
    ReactionRemovedEventSchema,
    DmReceivedEventSchema,
    GroupDmReceivedEventSchema,
    AgentOnlineEventSchema,
    AgentOfflineEventSchema,
    AgentSpawnRequestedEventSchema,
    AgentReleaseRequestedEventSchema,
    ChannelCreatedEventSchema,
    ChannelUpdatedEventSchema,
    ChannelArchivedEventSchema,
    MemberJoinedEventSchema,
    MemberLeftEventSchema,
    ChannelMutedEventSchema,
    ChannelUnmutedEventSchema,
    MessageReadEventSchema,
    FileUploadedEventSchema,
    WebhookReceivedEventSchema,
    CommandInvokedEventSchema,
    PongEventSchema,
]);
// Union of all events that WsClient can emit (includes server events + client-only events)
export const WsClientEventSchema = z.discriminatedUnion('type', [
    // Server events
    MessageCreatedEventSchema,
    MessageUpdatedEventSchema,
    ThreadReplyEventSchema,
    ReactionAddedEventSchema,
    ReactionRemovedEventSchema,
    DmReceivedEventSchema,
    GroupDmReceivedEventSchema,
    AgentOnlineEventSchema,
    AgentOfflineEventSchema,
    AgentSpawnRequestedEventSchema,
    AgentReleaseRequestedEventSchema,
    ChannelCreatedEventSchema,
    ChannelUpdatedEventSchema,
    ChannelArchivedEventSchema,
    MemberJoinedEventSchema,
    MemberLeftEventSchema,
    ChannelMutedEventSchema,
    ChannelUnmutedEventSchema,
    MessageReadEventSchema,
    FileUploadedEventSchema,
    WebhookReceivedEventSchema,
    CommandInvokedEventSchema,
    PongEventSchema,
    // Client-only events
    WsOpenEventSchema,
    WsErrorEventSchema,
    WsReconnectingEventSchema,
    WsPermanentlyDisconnectedEventSchema,
    WsCloseEventSchema,
]);
