// @relaycast/types/dist/workspace.js
import { z } from 'zod';
export const WorkspaceSchema = z.object({
    id: z.string(),
    name: z.string(),
    system_prompt: z.string().nullable(),
    plan: z.enum(['free', 'pro', 'enterprise']),
    created_at: z.string(),
    metadata: z.record(z.string(), z.unknown()),
});
export const CreateWorkspaceRequestSchema = z.object({
    name: z.string(),
});
export const CreateWorkspaceResponseSchema = z.object({
    workspace_id: z.string(),
    api_key: z.string(),
    created_at: z.string(),
});
export const WorkspaceLookupSchema = z.object({
    id: z.string(),
    name: z.string(),
    created_at: z.string(),
});
export const UpdateWorkspaceRequestSchema = z.object({
    name: z.string().optional(),
    system_prompt: z.string().nullable().optional(),
});
export const SystemPromptSchema = z.object({
    prompt: z.string(),
    is_default: z.boolean(),
});
export const SetSystemPromptRequestSchema = z.object({
    prompt: z.string().nullable().optional(),
    reset: z.boolean().optional(),
});
// --- Activity, DM conversations, token rotation ---
export const ActivityItemSchema = z.object({
    type: z.enum(['message', 'dm']),
    id: z.string(),
    channel_name: z.string().optional(),
    conversation_id: z.string().optional(),
    agent_name: z.string(),
    text: z.string(),
    created_at: z.string(),
});
export const WorkspaceDmConversationSchema = z.object({
    id: z.string(),
    channel_id: z.string(),
    type: z.string(),
    participants: z.array(z.string()),
    last_message: z.object({
        text: z.string(),
        agent_name: z.string(),
        created_at: z.string(),
    }).nullable(),
    message_count: z.number(),
});
export const TokenRotateResponseSchema = z.object({
    name: z.string(),
    token: z.string(),
});
