// @agent-relay/config/dist/bridge-config.js
/**
 * Bridge Configuration
 * Handles loading and resolving bridge configuration from files and CLI args.
 */
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { getProjectPaths } from './project-namespace.js';
const CONFIG_PATHS = [
    path.join(os.homedir(), '.agent-relay', 'bridge.json'),
    path.join(os.homedir(), '.config', 'agent-relay', 'bridge.json'),
];
/**
 * Load bridge config from file
 */
export function loadBridgeConfig() {
    for (const configPath of CONFIG_PATHS) {
        if (fs.existsSync(configPath)) {
            try {
                const content = fs.readFileSync(configPath, 'utf-8');
                if (typeof content !== 'string') {
                    continue;
                }
                const trimmed = content.trim();
                if (!trimmed) {
                    continue;
                }
                return JSON.parse(trimmed);
            }
            catch (err) {
                console.error(`[bridge] Failed to parse ${configPath}:`, err);
            }
        }
    }
    return null;
}
/**
 * Resolve project path (expand ~ and make absolute)
 */
export function resolvePath(p) {
    if (p.startsWith('~')) {
        p = path.join(os.homedir(), p.slice(1));
    }
    return path.resolve(p);
}
/**
 * Get default lead name from directory name
 */
export function getDefaultLeadName(projectPath) {
    const dirname = path.basename(projectPath);
    // Capitalize first letter
    return dirname.charAt(0).toUpperCase() + dirname.slice(1);
}
/**
 * Resolve projects from CLI args and/or config file
 */
export function resolveProjects(cliPaths, cliOverride) {
    const config = loadBridgeConfig();
    const projects = [];
    // If CLI paths provided, use those
    if (cliPaths.length > 0) {
        for (const p of cliPaths) {
            const projectPath = resolvePath(p);
            if (!fs.existsSync(projectPath)) {
                console.error(`[bridge] Project path does not exist: ${projectPath}`);
                continue;
            }
            const paths = getProjectPaths(projectPath);
            // Check for project-specific config
            const projectConfig = config?.projects?.[p] || config?.projects?.[projectPath];
            projects.push({
                path: projectPath,
                id: paths.projectId,
                socketPath: paths.socketPath,
                leadName: projectConfig?.lead || getDefaultLeadName(projectPath),
                cli: cliOverride || projectConfig?.cli || config?.defaultCli || 'claude',
            });
        }
    }
    // Otherwise use config file
    else if (config?.projects) {
        for (const [p, projectConfig] of Object.entries(config.projects)) {
            const projectPath = resolvePath(p);
            if (!fs.existsSync(projectPath)) {
                console.error(`[bridge] Project path does not exist: ${projectPath}`);
                continue;
            }
            const paths = getProjectPaths(projectPath);
            projects.push({
                path: projectPath,
                id: paths.projectId,
                socketPath: paths.socketPath,
                leadName: projectConfig.lead || getDefaultLeadName(projectPath),
                cli: cliOverride || projectConfig.cli || config.defaultCli || 'claude',
            });
        }
    }
    return projects;
}
/**
 * Validate that brokers are running for all projects
 */
export function validateBrokers(projects) {
    const valid = [];
    const missing = [];
    for (const project of projects) {
        if (fs.existsSync(project.socketPath)) {
            valid.push(project);
        }
        else {
            missing.push(project);
        }
    }
    return { valid, missing };
}
/**
 * Start brokers for missing projects
 */
export async function startMissingBrokers(projects) {
    const { execAsync } = await import('./bridge-utils.js');
    for (const project of projects) {
        console.log(`[bridge] Starting broker for ${project.id}...`);
        try {
            // Start broker in background
            await execAsync(`cd "${project.path}" && agent-relay up &`, {
                timeout: 5000,
            });
            // Wait for socket to appear
            await waitForSocket(project.socketPath, 10000);
            console.log(`[bridge] Broker started for ${project.id}`);
        }
        catch (err) {
            console.error(`[bridge] Failed to start broker for ${project.id}:`, err);
        }
    }
}
/**
 * Wait for socket file to exist
 */
async function waitForSocket(socketPath, timeoutMs) {
    const startTime = Date.now();
    while (Date.now() - startTime < timeoutMs) {
        if (fs.existsSync(socketPath)) {
            return;
        }
        await new Promise(r => setTimeout(r, 200));
    }
    throw new Error(`Timeout waiting for socket: ${socketPath}`);
}
