// @agent-relay/config/dist/cloud-config.js
/**
 * Agent Relay Cloud - Configuration
 */
function requireEnv(name) {
    const value = process.env[name];
    if (!value) {
        throw new Error(`Missing required environment variable: ${name}`);
    }
    return value;
}
function optionalEnv(name) {
    return process.env[name];
}
export function loadConfig() {
    return {
        port: parseInt(process.env.PORT || '4567', 10),
        publicUrl: process.env.PUBLIC_URL || 'http://localhost:4567',
        appUrl: process.env.APP_URL || process.env.PUBLIC_URL || 'http://localhost:4567',
        sessionSecret: requireEnv('SESSION_SECRET'),
        // Local dashboard for channel API (auto-detected if not set)
        localDashboardUrl: optionalEnv('LOCAL_DASHBOARD_URL'),
        databaseUrl: requireEnv('DATABASE_URL'),
        redisUrl: process.env.REDIS_URL || 'redis://localhost:6379',
        github: {
            clientId: requireEnv('GITHUB_CLIENT_ID'),
            clientSecret: requireEnv('GITHUB_CLIENT_SECRET'),
            webhookSecret: optionalEnv('GITHUB_WEBHOOK_SECRET'),
        },
        providers: {
            anthropic: optionalEnv('ANTHROPIC_CLIENT_ID')
                ? { clientId: optionalEnv('ANTHROPIC_CLIENT_ID') }
                : undefined,
            openai: optionalEnv('OPENAI_CLIENT_ID')
                ? { clientId: optionalEnv('OPENAI_CLIENT_ID') }
                : undefined,
            google: optionalEnv('GOOGLE_CLIENT_ID') && optionalEnv('GOOGLE_CLIENT_SECRET')
                ? {
                    clientId: optionalEnv('GOOGLE_CLIENT_ID'),
                    clientSecret: optionalEnv('GOOGLE_CLIENT_SECRET'),
                }
                : undefined,
        },
        vault: {
            masterKey: requireEnv('VAULT_MASTER_KEY'),
        },
        compute: {
            provider: process.env.COMPUTE_PROVIDER || 'docker',
            fly: optionalEnv('FLY_API_TOKEN')
                ? {
                    apiToken: optionalEnv('FLY_API_TOKEN'),
                    org: optionalEnv('FLY_ORG') || 'personal',
                    region: optionalEnv('FLY_REGION') || 'sjc',
                    workspaceDomain: optionalEnv('FLY_WORKSPACE_DOMAIN'),
                    registryAuth: optionalEnv('GHCR_USERNAME') && optionalEnv('GHCR_TOKEN')
                        ? {
                            username: optionalEnv('GHCR_USERNAME'),
                            password: optionalEnv('GHCR_TOKEN'),
                        }
                        : undefined,
                    snapshotRetentionDays: parseInt(optionalEnv('FLY_SNAPSHOT_RETENTION_DAYS') || '14', 10),
                    volumeSizeGb: parseInt(optionalEnv('FLY_VOLUME_SIZE_GB') || '10', 10),
                }
                : undefined,
            railway: optionalEnv('RAILWAY_API_TOKEN')
                ? {
                    apiToken: optionalEnv('RAILWAY_API_TOKEN'),
                }
                : undefined,
        },
        nango: {
            secretKey: requireEnv('NANGO_SECRET_KEY'),
            host: optionalEnv('NANGO_HOST'),
        },
        stripe: {
            secretKey: requireEnv('STRIPE_SECRET_KEY'),
            publishableKey: requireEnv('STRIPE_PUBLISHABLE_KEY'),
            webhookSecret: requireEnv('STRIPE_WEBHOOK_SECRET'),
            priceIds: {
                proMonthly: optionalEnv('STRIPE_PRO_MONTHLY_PRICE_ID'),
                proYearly: optionalEnv('STRIPE_PRO_YEARLY_PRICE_ID'),
                teamMonthly: optionalEnv('STRIPE_TEAM_MONTHLY_PRICE_ID'),
                teamYearly: optionalEnv('STRIPE_TEAM_YEARLY_PRICE_ID'),
                enterpriseMonthly: optionalEnv('STRIPE_ENTERPRISE_MONTHLY_PRICE_ID'),
                enterpriseYearly: optionalEnv('STRIPE_ENTERPRISE_YEARLY_PRICE_ID'),
            },
        },
        // Admin users - comma-separated GitHub usernames (e.g., "khaliqgant,admin2")
        adminUsers: (optionalEnv('ADMIN_USERS') || '')
            .split(',')
            .map((u) => u.trim().toLowerCase())
            .filter(Boolean),
    };
}
/**
 * Check if a GitHub username is an admin user
 */
export function isAdminUser(githubUsername) {
    const config = getConfig();
    return config.adminUsers.includes(githubUsername.toLowerCase());
}
// Singleton config instance
let _config = null;
export function getConfig() {
    if (!_config) {
        _config = loadConfig();
    }
    return _config;
}
