// @agent-relay/sdk/dist/communicate/a2a-types.js
/**
 * A2A (Agent-to-Agent) protocol data model types.
 */
import { randomUUID } from 'node:crypto';
export const VALID_TASK_STATES = new Set([
    'submitted',
    'working',
    'completed',
    'failed',
    'canceled',
]);
// --- Factories ---
export function createA2APart(text) {
    const part = {};
    if (text !== undefined)
        part.text = text;
    return part;
}
export function createA2AMessage(role, parts, opts) {
    return {
        role,
        parts,
        messageId: opts?.messageId ?? randomUUID(),
        contextId: opts?.contextId,
        taskId: opts?.taskId,
    };
}
export function createA2ATaskStatus(state, message) {
    return {
        state,
        message,
        timestamp: new Date().toISOString(),
    };
}
export function createA2ATask(id, contextId) {
    return {
        id,
        contextId,
        status: createA2ATaskStatus('submitted'),
        messages: [],
        artifacts: [],
    };
}
export function createA2AAgentCard(name, description, url, skills = []) {
    return {
        name,
        description,
        url,
        version: '1.0.0',
        capabilities: { streaming: true, pushNotifications: false },
        skills,
        defaultInputModes: ['text'],
        defaultOutputModes: ['text'],
    };
}
// --- Serialization helpers ---
export function a2aPartToDict(part) {
    const d = {};
    if (part.text !== undefined)
        d.text = part.text;
    if (part.file !== undefined)
        d.file = part.file;
    if (part.data !== undefined)
        d.data = part.data;
    return d;
}
export function a2aPartFromDict(d) {
    return {
        text: d.text,
        file: d.file,
        data: d.data,
    };
}
export function a2aMessageToDict(msg) {
    const d = {
        role: msg.role,
        parts: msg.parts.map(a2aPartToDict),
    };
    if (msg.messageId !== undefined)
        d.messageId = msg.messageId;
    if (msg.contextId !== undefined)
        d.contextId = msg.contextId;
    if (msg.taskId !== undefined)
        d.taskId = msg.taskId;
    return d;
}
export function a2aMessageFromDict(d) {
    const parts = (d.parts ?? []).map(a2aPartFromDict);
    return {
        role: d.role,
        parts,
        messageId: d.messageId,
        contextId: d.contextId,
        taskId: d.taskId,
    };
}
export function a2aMessageGetText(msg) {
    return msg.parts
        .filter((p) => p.text)
        .map((p) => p.text)
        .join(' ');
}
export function a2aTaskStatusToDict(status) {
    const d = { state: status.state };
    if (status.message !== undefined)
        d.message = a2aMessageToDict(status.message);
    if (status.timestamp !== undefined)
        d.timestamp = status.timestamp;
    return d;
}
export function a2aTaskStatusFromDict(d) {
    let message;
    if (d.message !== undefined && d.message !== null) {
        message = a2aMessageFromDict(d.message);
    }
    return {
        state: d.state,
        message,
        timestamp: d.timestamp,
    };
}
export function a2aTaskToDict(task) {
    return {
        id: task.id,
        contextId: task.contextId,
        status: a2aTaskStatusToDict(task.status),
        messages: task.messages.map(a2aMessageToDict),
        artifacts: task.artifacts,
    };
}
export function a2aTaskFromDict(d) {
    const status = d.status
        ? a2aTaskStatusFromDict(d.status)
        : createA2ATaskStatus('submitted');
    const messages = (d.messages ?? []).map(a2aMessageFromDict);
    return {
        id: d.id,
        contextId: d.contextId,
        status,
        messages,
        artifacts: d.artifacts ?? [],
    };
}
export function a2aSkillToDict(skill) {
    return { id: skill.id, name: skill.name, description: skill.description };
}
export function a2aSkillFromDict(d) {
    return {
        id: d.id,
        name: d.name,
        description: d.description,
    };
}
export function a2aAgentCardToDict(card) {
    return {
        name: card.name,
        description: card.description,
        url: card.url,
        version: card.version,
        capabilities: card.capabilities,
        skills: card.skills.map(a2aSkillToDict),
        defaultInputModes: card.defaultInputModes,
        defaultOutputModes: card.defaultOutputModes,
    };
}
export function a2aAgentCardFromDict(d) {
    const skills = (d.skills ?? []).map(a2aSkillFromDict);
    return {
        name: d.name,
        description: d.description,
        url: d.url,
        version: d.version ?? '1.0.0',
        capabilities: d.capabilities ?? {
            streaming: true,
            pushNotifications: false,
        },
        skills,
        defaultInputModes: d.defaultInputModes ?? ['text'],
        defaultOutputModes: d.defaultOutputModes ?? ['text'],
    };
}
export function makeJsonRpcRequest(method, params, id) {
    return {
        jsonrpc: '2.0',
        method,
        params,
        id: id ?? randomUUID(),
    };
}
export function makeJsonRpcResponse(result, id) {
    return { jsonrpc: '2.0', result, id };
}
export function makeJsonRpcError(code, message, id) {
    return {
        jsonrpc: '2.0',
        error: { code, message },
        id,
    };
}
// Standard JSON-RPC error codes
export const JSONRPC_PARSE_ERROR = -32700;
export const JSONRPC_INVALID_REQUEST = -32600;
export const JSONRPC_METHOD_NOT_FOUND = -32601;
export const JSONRPC_INVALID_PARAMS = -32602;
export const JSONRPC_INTERNAL_ERROR = -32603;
// A2A-specific error codes
export const A2A_TASK_NOT_FOUND = -32001;
export const A2A_TASK_NOT_CANCELABLE = -32002;
