// @agent-relay/sdk/dist/communicate/a2a-bridge.js
/**
 * Bridge that connects an external A2A agent into a Relay workspace.
 *
 * - Registers a proxy agent on the Relay workspace
 * - When Relay messages arrive for the proxy, forwards them as A2A JSON-RPC
 *   message/send calls to the external agent
 * - When A2A responses come back, forwards them as Relay DMs
 */
import { randomUUID } from 'node:crypto';
import { Relay } from './core.js';
import { a2aAgentCardFromDict, } from './a2a-types.js';
export class A2ABridge {
    relay;
    a2aAgentUrl;
    proxyName;
    _agentCard;
    _started = false;
    constructor(relayConfig, a2aAgentUrl, proxyName) {
        this.relay = new Relay(proxyName, relayConfig);
        this.a2aAgentUrl = a2aAgentUrl.replace(/\/+$/, '');
        this.proxyName = proxyName;
    }
    async start() {
        this.relay.onMessage((msg) => this._handleRelayMessage(msg));
        this._started = true;
    }
    async stop() {
        this._started = false;
        await this.relay.close();
    }
    async discoverAgent() {
        const url = `${this.a2aAgentUrl}/.well-known/agent.json`;
        const response = await fetch(url);
        const data = (await response.json());
        this._agentCard = a2aAgentCardFromDict(data);
        return this._agentCard;
    }
    async sendA2AMessage(text) {
        const a2aMsg = {
            role: 'user',
            parts: [{ text }],
            messageId: randomUUID(),
        };
        const jsonrpcRequest = {
            jsonrpc: '2.0',
            method: 'message/send',
            params: { message: a2aMsg },
            id: randomUUID(),
        };
        const targetUrl = this._agentCard?.url ?? this.a2aAgentUrl;
        const response = await fetch(targetUrl, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(jsonrpcRequest),
        });
        const data = (await response.json());
        const result = (data.result ?? {});
        // Extract response text from task status message
        const status = (result.status ?? {});
        const statusMsg = (status.message ?? {});
        const statusParts = (statusMsg.parts ?? []);
        if (statusParts.length > 0) {
            const t = statusParts[0].text;
            if (t)
                return t;
        }
        // Try from messages list
        const messages = (result.messages ?? []);
        for (let i = messages.length - 1; i >= 0; i--) {
            const msg = messages[i];
            if (msg.role === 'agent') {
                const msgParts = (msg.parts ?? []);
                if (msgParts.length > 0) {
                    const t = msgParts[0].text;
                    if (t)
                        return t;
                }
            }
        }
        return null;
    }
    async _handleRelayMessage(msg) {
        const responseText = await this.sendA2AMessage(msg.text);
        if (responseText) {
            await this.relay.send(msg.sender, responseText);
        }
    }
}
