// @agent-relay/sdk/dist/communicate/a2a-server.js
/**
 * A2A-compliant HTTP server that exposes a Relay agent as an A2A endpoint.
 *
 * Routes:
 *   GET  /.well-known/agent.json  -> Agent Card
 *   POST /                        -> JSON-RPC 2.0 dispatcher
 */
import { randomUUID } from 'node:crypto';
import * as http from 'node:http';
import { a2aAgentCardToDict, a2aMessageToDict, createA2AAgentCard, createA2ATaskStatus, } from './a2a-types.js';
export class A2AServer {
    agentName;
    port;
    host;
    skills;
    tasks = new Map();
    _onMessage;
    _server;
    _actualPort;
    constructor(agentName, port = 5000, host = '0.0.0.0', skills = []) {
        this.agentName = agentName;
        this.port = port;
        this.host = host;
        this.skills = skills;
    }
    get url() {
        const p = this._actualPort ?? this.port;
        return `http://${this.host}:${p}`;
    }
    onMessage(callback) {
        this._onMessage = callback;
    }
    getAgentCard() {
        return createA2AAgentCard(this.agentName, `Agent Relay agent: ${this.agentName}`, this.url, [...this.skills]);
    }
    async handleMessageSend(params) {
        const messageData = (params.message ?? {});
        const rawParts = (messageData.parts ?? []);
        const parts = rawParts.map((p) => ({ text: p.text }));
        const incoming = {
            role: messageData.role ?? 'user',
            parts,
            messageId: messageData.messageId ?? randomUUID(),
            contextId: messageData.contextId,
            taskId: messageData.taskId,
        };
        // Create or find task
        const taskId = incoming.taskId ?? randomUUID();
        const contextId = incoming.contextId ?? randomUUID();
        let task;
        if (this.tasks.has(taskId)) {
            task = this.tasks.get(taskId);
            task.messages.push(incoming);
            task.status = createA2ATaskStatus('working');
        }
        else {
            task = {
                id: taskId,
                contextId,
                status: createA2ATaskStatus('working'),
                messages: [incoming],
                artifacts: [],
            };
            this.tasks.set(taskId, task);
        }
        // Invoke callback
        let responseMsg = null;
        if (this._onMessage) {
            const result = this._onMessage(incoming);
            responseMsg = result instanceof Promise ? await result : result;
        }
        if (responseMsg) {
            task.messages.push(responseMsg);
            task.status = createA2ATaskStatus('completed', responseMsg);
        }
        else {
            task.status = createA2ATaskStatus('completed');
        }
        return taskToDict(task);
    }
    async handleTasksGet(taskId) {
        const task = this.tasks.get(taskId);
        if (!task) {
            throw new Error(`Task not found: ${taskId}`);
        }
        return taskToDict(task);
    }
    async handleTasksCancel(taskId) {
        const task = this.tasks.get(taskId);
        if (!task) {
            throw new Error(`Task not found: ${taskId}`);
        }
        task.status = createA2ATaskStatus('canceled');
        return taskToDict(task);
    }
    async start() {
        return new Promise((resolve, reject) => {
            const server = http.createServer((req, res) => {
                void this._handleRequest(req, res);
            });
            server.on('error', reject);
            server.listen(this.port, this.host, () => {
                this._server = server;
                const addr = server.address();
                if (typeof addr === 'object' && addr) {
                    this._actualPort = addr.port;
                }
                resolve();
            });
        });
    }
    async stop() {
        if (this._server) {
            await new Promise((resolve, reject) => {
                this._server.close((err) => (err ? reject(err) : resolve()));
            });
            this._server = undefined;
            this._actualPort = undefined;
        }
    }
    // --- HTTP handlers ---
    async _handleRequest(req, res) {
        if (req.method === 'GET' && req.url === '/.well-known/agent.json') {
            const card = this.getAgentCard();
            res.writeHead(200, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify(a2aAgentCardToDict(card)));
            return;
        }
        if (req.method === 'POST' && (req.url === '/' || req.url === '')) {
            await this._handleJsonRpc(req, res);
            return;
        }
        res.writeHead(404, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ error: 'Not found' }));
    }
    async _handleJsonRpc(req, res) {
        let body;
        try {
            const raw = await readBody(req);
            body = JSON.parse(raw);
        }
        catch {
            res.writeHead(400, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({
                jsonrpc: '2.0',
                error: { code: -32700, message: 'Parse error' },
                id: null,
            }));
            return;
        }
        const method = (body.method ?? '');
        const params = (body.params ?? {});
        const rpcId = body.id ?? null;
        try {
            let result;
            if (method === 'message/send') {
                result = await this.handleMessageSend(params);
            }
            else if (method === 'tasks/get') {
                const id = (params.id ?? params.taskId ?? '');
                result = await this.handleTasksGet(id);
            }
            else if (method === 'tasks/cancel') {
                const id = (params.id ?? params.taskId ?? '');
                result = await this.handleTasksCancel(id);
            }
            else {
                res.writeHead(400, { 'Content-Type': 'application/json' });
                res.end(JSON.stringify({
                    jsonrpc: '2.0',
                    error: { code: -32601, message: `Method not found: ${method}` },
                    id: rpcId,
                }));
                return;
            }
            res.writeHead(200, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ jsonrpc: '2.0', result, id: rpcId }));
        }
        catch (err) {
            res.writeHead(404, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({
                jsonrpc: '2.0',
                error: { code: -32602, message: String(err) },
                id: rpcId,
            }));
        }
    }
}
// --- Helpers ---
function readBody(req) {
    return new Promise((resolve, reject) => {
        const chunks = [];
        req.on('data', (chunk) => chunks.push(chunk));
        req.on('end', () => resolve(Buffer.concat(chunks).toString()));
        req.on('error', reject);
    });
}
function taskToDict(task) {
    const statusDict = { state: task.status.state };
    if (task.status.message) {
        statusDict.message = a2aMessageToDict(task.status.message);
    }
    const messages = task.messages.map((m) => {
        const md = {
            role: m.role,
            parts: m.parts.map((p) => ({ text: p.text })),
        };
        if (m.messageId)
            md.messageId = m.messageId;
        return md;
    });
    return {
        id: task.id,
        contextId: task.contextId,
        status: statusDict,
        messages,
        artifacts: task.artifacts,
    };
}
