// @agent-relay/sdk/dist/communicate/transport.js
import { setTimeout as sleep } from 'node:timers/promises';
import { WebSocket } from 'ws';
import { RelayAuthError, RelayConfigError, RelayConnectionError, resolveRelayConfig, } from './types.js';
const HTTP_RETRY_ATTEMPTS = 3;
const WS_RECONNECT_MAX_DELAY_MS = 30_000;
export class RelayTransport {
    agentName;
    config;
    agentId;
    token;
    ws;
    messageCallback;
    closing = false;
    reconnectDelayMs = 1_000;
    reconnectTimer;
    wsConnectPromise;
    constructor(agentName, config = {}) {
        this.agentName = agentName;
        this.config = resolveRelayConfig(config);
    }
    async connect() {
        this.requireConfig({ requireWorkspace: true });
        this.closing = false;
        await this.registerAgent();
        await this.connectWebSocket();
    }
    async disconnect() {
        this.closing = true;
        this.clearReconnectTimer();
        const socket = this.ws;
        this.ws = undefined;
        this.wsConnectPromise = undefined;
        if (socket && socket.readyState !== WebSocket.CLOSED) {
            await Promise.race([
                new Promise((resolve) => {
                    socket.once('close', () => resolve());
                    socket.close();
                }),
                new Promise((resolve) => setTimeout(resolve, 2_000)),
            ]);
        }
        if (this.agentId) {
            try {
                await this.unregisterAgent();
            }
            catch {
                // Best-effort cleanup.
            }
        }
        this.closing = false;
    }
    onWsMessage(callback) {
        this.messageCallback = callback;
    }
    async registerAgent() {
        this.requireConfig({ requireWorkspace: true });
        if (this.agentId && this.token) {
            return this.agentId;
        }
        const payload = await this.sendHttp('POST', '/v1/agents/register', {
            name: this.agentName,
            workspace: this.config.workspace,
        });
        this.agentId = payload.agent_id;
        this.token = payload.token;
        return this.agentId;
    }
    async unregisterAgent() {
        if (!this.agentId) {
            return;
        }
        const agentId = this.agentId;
        this.agentId = undefined;
        this.token = undefined;
        await this.sendHttp('DELETE', `/v1/agents/${agentId}`);
    }
    async sendDm(to, text) {
        await this.ensureRegistered();
        const payload = await this.sendHttp('POST', '/v1/messages/dm', {
            to,
            text,
            from: this.agentName,
        });
        return payload.message_id;
    }
    async postMessage(channel, text) {
        await this.ensureRegistered();
        const payload = await this.sendHttp('POST', '/v1/messages/channel', {
            channel,
            text,
            from: this.agentName,
        });
        return payload.message_id;
    }
    async reply(messageId, text) {
        await this.ensureRegistered();
        const payload = await this.sendHttp('POST', '/v1/messages/reply', {
            message_id: messageId,
            text,
            from: this.agentName,
        });
        return payload.message_id;
    }
    async checkInbox() {
        await this.ensureRegistered();
        const payload = await this.sendHttp('GET', `/v1/inbox/${this.agentId}`);
        return (payload.messages ?? []).map((message) => this.messageFromPayload(message));
    }
    async listAgents() {
        const payload = await this.sendHttp('GET', '/v1/agents');
        return [...(payload.agents ?? [])];
    }
    async ensureRegistered() {
        if (this.agentId && this.token) {
            return;
        }
        await this.registerAgent();
    }
    requireConfig(options = {}) {
        if (!this.config.apiKey) {
            throw new RelayConfigError('Missing RELAY_API_KEY. Set the environment variable or pass apiKey to RelayConfig.');
        }
        if (options.requireWorkspace && !this.config.workspace) {
            throw new RelayConfigError('Missing RELAY_WORKSPACE. Set the environment variable or pass workspace to RelayConfig.');
        }
    }
    async sendHttp(method, path, payload) {
        this.requireConfig();
        const url = `${this.config.baseUrl}${path}`;
        const headers = {
            authorization: `Bearer ${this.config.apiKey}`,
        };
        if (payload !== undefined) {
            headers['content-type'] = 'application/json';
        }
        for (let attempt = 1; attempt <= HTTP_RETRY_ATTEMPTS; attempt += 1) {
            let response;
            try {
                response = await fetch(url, {
                    method,
                    headers,
                    body: payload === undefined ? undefined : JSON.stringify(payload),
                });
            }
            catch (error) {
                if (attempt < HTTP_RETRY_ATTEMPTS) {
                    await sleep(Math.min(2 ** (attempt - 1) * 1_000, WS_RECONNECT_MAX_DELAY_MS));
                    continue;
                }
                throw new RelayConnectionError(0, error instanceof Error ? error.message : String(error));
            }
            if (response.status === 401) {
                throw new RelayAuthError(await this.errorMessage(response));
            }
            if (response.status >= 500 && response.status <= 599) {
                const message = await this.errorMessage(response);
                if (attempt < HTTP_RETRY_ATTEMPTS) {
                    await sleep(Math.min(2 ** (attempt - 1) * 1_000, WS_RECONNECT_MAX_DELAY_MS));
                    continue;
                }
                throw new RelayConnectionError(response.status, message);
            }
            if (response.status >= 400) {
                throw new RelayConnectionError(response.status, await this.errorMessage(response));
            }
            if (response.status === 204) {
                return undefined;
            }
            const contentType = response.headers.get('content-type') ?? '';
            if (contentType.includes('application/json')) {
                return (await response.json());
            }
            return (await response.text());
        }
        throw new RelayConnectionError(500, 'Unexpected transport retry failure');
    }
    async connectWebSocket() {
        await this.ensureRegistered();
        if (this.ws && (this.ws.readyState === WebSocket.OPEN || this.ws.readyState === WebSocket.CONNECTING)) {
            return;
        }
        if (this.wsConnectPromise) {
            return this.wsConnectPromise;
        }
        const url = `${this.wsBaseUrl()}/v1/ws/${this.agentId}?token=${encodeURIComponent(this.token ?? '')}`;
        const socket = new WebSocket(url);
        this.wsConnectPromise = new Promise((resolve, reject) => {
            let settled = false;
            socket.once('open', () => {
                settled = true;
                this.ws = socket;
                this.reconnectDelayMs = 1_000;
                resolve();
            });
            socket.once('error', (error) => {
                if (!settled) {
                    settled = true;
                    reject(error);
                }
            });
        }).finally(() => {
            this.wsConnectPromise = undefined;
        });
        socket.on('message', (raw) => {
            void this.dispatchWsPayload(raw.toString());
        });
        socket.on('close', () => {
            if (this.ws === socket) {
                this.ws = undefined;
            }
            if (!this.closing) {
                this.scheduleReconnect();
            }
        });
        socket.on('error', () => {
            // The close handler manages reconnects for established sockets.
        });
        await this.wsConnectPromise;
    }
    wsBaseUrl() {
        if (this.config.baseUrl.startsWith('https://')) {
            return `wss://${this.config.baseUrl.slice('https://'.length)}`;
        }
        if (this.config.baseUrl.startsWith('http://')) {
            return `ws://${this.config.baseUrl.slice('http://'.length)}`;
        }
        return this.config.baseUrl;
    }
    scheduleReconnect() {
        if (this.reconnectTimer || this.closing) {
            return;
        }
        const delayMs = this.reconnectDelayMs;
        this.reconnectDelayMs = Math.min(this.reconnectDelayMs * 2, WS_RECONNECT_MAX_DELAY_MS);
        this.reconnectTimer = setTimeout(() => {
            this.reconnectTimer = undefined;
            void this.connectWebSocket().catch(() => {
                this.scheduleReconnect();
            });
        }, delayMs);
    }
    clearReconnectTimer() {
        if (this.reconnectTimer) {
            clearTimeout(this.reconnectTimer);
            this.reconnectTimer = undefined;
        }
    }
    async dispatchWsPayload(rawPayload) {
        const payload = JSON.parse(rawPayload);
        if (payload.type === 'ping') {
            this.ws?.send(JSON.stringify({ type: 'pong' }));
            return;
        }
        if (payload.type !== 'message' || !this.messageCallback) {
            return;
        }
        await this.messageCallback(this.messageFromPayload(payload));
    }
    messageFromPayload(payload) {
        return {
            sender: String(payload.sender ?? ''),
            text: String(payload.text ?? ''),
            channel: typeof payload.channel === 'string' ? payload.channel : undefined,
            threadId: typeof payload.thread_id === 'string' ? payload.thread_id : undefined,
            timestamp: typeof payload.timestamp === 'number' ? payload.timestamp : undefined,
            messageId: typeof payload.message_id === 'string' ? payload.message_id : undefined,
        };
    }
    async errorMessage(response) {
        try {
            const payload = (await response.json());
            return payload.message ?? response.statusText ?? 'Request failed';
        }
        catch {
            return (await response.text()) || response.statusText || 'Request failed';
        }
    }
}
