// @agent-relay/sdk/dist/communicate/types.js
export const DEFAULT_RELAY_BASE_URL = 'https://api.relaycast.dev';
export class RelayConnectionError extends Error {
    statusCode;
    constructor(statusCode, message) {
        super(`${statusCode}: ${message}`);
        this.name = 'RelayConnectionError';
        this.statusCode = statusCode;
    }
}
export class RelayConfigError extends Error {
    constructor(message) {
        super(message);
        this.name = 'RelayConfigError';
    }
}
export class RelayAuthError extends RelayConnectionError {
    constructor(message = 'Unauthorized', statusCode = 401) {
        super(statusCode, message);
        this.name = 'RelayAuthError';
    }
}
/** Strip trailing slashes without a quantified regex (avoids ReDoS). */
function trimTrailingSlashes(url) {
    let end = url.length;
    while (end > 0 && url[end - 1] === '/') {
        end -= 1;
    }
    return url.slice(0, end);
}
/**
 * Resolve a partial config into a fully-populated config with env-var fallbacks.
 * @param config - Partial user config.
 * @returns Resolved config with all defaults applied.
 */
export function resolveRelayConfig(config = {}) {
    return {
        workspace: config.workspace ?? process.env.RELAY_WORKSPACE,
        apiKey: config.apiKey ?? process.env.RELAY_API_KEY,
        baseUrl: trimTrailingSlashes(config.baseUrl ?? process.env.RELAY_BASE_URL ?? DEFAULT_RELAY_BASE_URL),
        channels: [...(config.channels ?? ['general'])],
        pollIntervalMs: config.pollIntervalMs ?? 1_000,
        autoCleanup: config.autoCleanup ?? true,
    };
}
/**
 * Format a single message for display in agent instructions.
 * @param message - The relay message to format.
 * @returns Human-readable message string.
 */
export function formatRelayMessage(message) {
    const location = message.channel ? ` [#${message.channel}]` : '';
    const thread = message.threadId ? ` [thread ${message.threadId}]` : '';
    return `Relay message from ${message.sender}${location}${thread}: ${message.text}`;
}
/**
 * Format an array of messages for display in agent instructions.
 * @param messages - The relay messages to format.
 * @returns Human-readable inbox summary.
 */
export function formatRelayInbox(messages) {
    if (messages.length === 0) {
        return 'No new relay messages.';
    }
    return messages.map((message) => formatRelayMessage(message)).join('\n');
}
