// @agent-relay/sdk/dist/consensus.js
/**
 * Consensus / voting engine for distributed agent decision-making.
 *
 * Runs entirely in the SDK process — no broker protocol extension needed.
 * Proposals are broadcast and votes collected via the relay messaging layer.
 *
 * Consensus types:
 * - majority   — simple >50%
 * - supermajority — configurable threshold (default 2/3)
 * - unanimous  — all participants must approve
 * - weighted   — votes weighted by agent role/expertise
 * - quorum     — minimum participation + majority
 */
import { randomBytes } from "node:crypto";
import { EventEmitter } from "node:events";
export { formatProposalMessage, formatResultMessage, parseVoteCommand, isConsensusCommand, parseProposalCommand, } from "./consensus-helpers.js";
// ── Defaults ─────────────────────────────────────────────────────────────────
const DEFAULT_CONFIG = {
    defaultTimeoutMs: 5 * 60 * 1000,
    defaultConsensusType: "majority",
    defaultThreshold: 0.67,
    allowVoteChange: true,
    autoResolve: true,
    maxRetainedProposals: 100,
};
// ── Engine ───────────────────────────────────────────────────────────────────
export class ConsensusEngine extends EventEmitter {
    config;
    proposals = new Map();
    expiryTimers = new Map();
    constructor(config = {}) {
        super();
        this.config = { ...DEFAULT_CONFIG, ...config };
    }
    // ── Proposal management ──────────────────────────────────────────────────
    createProposal(options) {
        const id = `prop_${Date.now()}_${randomBytes(4).toString("hex")}`;
        const now = Date.now();
        const timeoutMs = options.timeoutMs ?? this.config.defaultTimeoutMs;
        const proposal = {
            id,
            title: options.title,
            description: options.description,
            proposer: options.proposer,
            consensusType: options.consensusType ?? this.config.defaultConsensusType,
            participants: options.participants,
            quorum: options.quorum,
            threshold: options.threshold ?? this.config.defaultThreshold,
            weights: options.weights,
            createdAt: now,
            expiresAt: now + timeoutMs,
            status: "pending",
            votes: [],
            metadata: options.metadata,
            thread: options.thread ?? `consensus-${id}`,
        };
        this.proposals.set(id, proposal);
        this.scheduleExpiry(proposal);
        this.emit("proposal:created", proposal);
        return proposal;
    }
    vote(proposalId, agent, value, reason) {
        const proposal = this.proposals.get(proposalId);
        if (!proposal)
            return { success: false, error: "Proposal not found" };
        if (proposal.status !== "pending")
            return { success: false, error: `Proposal is ${proposal.status}` };
        if (!proposal.participants.includes(agent))
            return { success: false, error: "Agent not a participant" };
        if (Date.now() > proposal.expiresAt) {
            this.expireProposal(proposal);
            return { success: false, error: "Proposal has expired" };
        }
        const existingIdx = proposal.votes.findIndex((v) => v.agent === agent);
        if (existingIdx >= 0) {
            if (!this.config.allowVoteChange)
                return {
                    success: false,
                    error: "Vote already cast and changes not allowed",
                };
            proposal.votes.splice(existingIdx, 1);
        }
        const weight = this.getAgentWeight(proposal, agent);
        const vote = { agent, value, weight, reason, timestamp: Date.now() };
        proposal.votes.push(vote);
        this.emit("proposal:voted", proposal, vote);
        if (this.config.autoResolve) {
            const result = this.calculateResult(proposal);
            if (this.canResolveEarly(proposal, result)) {
                this.resolveProposal(proposal, result);
            }
        }
        return { success: true, proposal };
    }
    getProposal(proposalId) {
        return this.proposals.get(proposalId) ?? null;
    }
    getProposalsForAgent(agent) {
        const out = [];
        for (const p of this.proposals.values()) {
            if (p.proposer === agent || p.participants.includes(agent))
                out.push(p);
        }
        return out;
    }
    getPendingVotesForAgent(agent) {
        const out = [];
        for (const p of this.proposals.values()) {
            if (p.status !== "pending")
                continue;
            if (!p.participants.includes(agent))
                continue;
            if (p.votes.some((v) => v.agent === agent))
                continue;
            out.push(p);
        }
        return out;
    }
    cancelProposal(proposalId, agent) {
        const proposal = this.proposals.get(proposalId);
        if (!proposal)
            return { success: false, error: "Proposal not found" };
        if (proposal.proposer !== agent)
            return { success: false, error: "Only proposer can cancel" };
        if (proposal.status !== "pending")
            return { success: false, error: `Proposal is ${proposal.status}` };
        proposal.status = "cancelled";
        this.clearExpiryTimer(proposalId);
        this.emit("proposal:cancelled", proposal);
        this.evictOldProposals();
        return { success: true };
    }
    forceResolve(proposalId) {
        const proposal = this.proposals.get(proposalId);
        if (!proposal || proposal.status !== "pending")
            return null;
        const result = this.calculateResult(proposal);
        this.resolveProposal(proposal, result);
        return result;
    }
    // ── Calculation ──────────────────────────────────────────────────────────
    calculateResult(proposal) {
        let approveWeight = 0;
        let rejectWeight = 0;
        let abstainWeight = 0;
        for (const v of proposal.votes) {
            if (v.value === "approve")
                approveWeight += v.weight;
            else if (v.value === "reject")
                rejectWeight += v.weight;
            else
                abstainWeight += v.weight;
        }
        const totalWeight = this.getTotalWeight(proposal);
        const votedWeight = approveWeight + rejectWeight + abstainWeight;
        const participation = totalWeight > 0 ? votedWeight / totalWeight : 0;
        const voters = new Set(proposal.votes.map((v) => v.agent));
        const nonVoters = proposal.participants.filter((p) => !voters.has(p));
        const quorumRequired = proposal.quorum ?? Math.ceil(proposal.participants.length / 2);
        const quorumMet = proposal.votes.length >= quorumRequired;
        const decision = this.determineDecision(proposal, {
            approveWeight,
            rejectWeight,
            votedWeight,
            quorumMet,
        });
        return {
            decision,
            approveWeight,
            rejectWeight,
            abstainWeight,
            participation,
            quorumMet,
            resolvedAt: Date.now(),
            nonVoters,
        };
    }
    determineDecision(proposal, counts) {
        const { approveWeight, rejectWeight, votedWeight, quorumMet } = counts;
        switch (proposal.consensusType) {
            case "unanimous": {
                if (proposal.votes.some((v) => v.value === "reject"))
                    return "rejected";
                if (proposal.votes.length < proposal.participants.length)
                    return "no_consensus";
                return proposal.votes.every((v) => v.value === "approve")
                    ? "approved"
                    : "rejected";
            }
            case "supermajority": {
                const threshold = proposal.threshold ?? this.config.defaultThreshold;
                if (votedWeight === 0)
                    return "no_consensus";
                if (approveWeight / votedWeight >= threshold)
                    return "approved";
                if (rejectWeight / votedWeight > 1 - threshold)
                    return "rejected";
                return "no_consensus";
            }
            case "quorum": {
                if (!quorumMet)
                    return "no_consensus";
                // fall through to majority
            }
            // eslint-disable-next-line no-fallthrough
            case "majority": {
                if (votedWeight === 0)
                    return "no_consensus";
                if (approveWeight > rejectWeight)
                    return "approved";
                if (rejectWeight > approveWeight)
                    return "rejected";
                return "no_consensus";
            }
            case "weighted": {
                if (votedWeight === 0)
                    return "no_consensus";
                if (approveWeight > rejectWeight)
                    return "approved";
                if (rejectWeight > approveWeight)
                    return "rejected";
                return "no_consensus";
            }
            default:
                return "no_consensus";
        }
    }
    canResolveEarly(proposal, result) {
        const totalWeight = this.getTotalWeight(proposal);
        const remainingWeight = totalWeight -
            (result.approveWeight + result.rejectWeight + result.abstainWeight);
        switch (proposal.consensusType) {
            case "unanimous":
                return (proposal.votes.some((v) => v.value === "reject") ||
                    proposal.votes.length === proposal.participants.length);
            case "supermajority": {
                const threshold = proposal.threshold ?? this.config.defaultThreshold;
                const votedWeight = result.approveWeight + result.rejectWeight + result.abstainWeight;
                if (votedWeight > 0 &&
                    result.approveWeight / votedWeight >= threshold) {
                    return (result.approveWeight / (votedWeight + remainingWeight) >= threshold);
                }
                if (votedWeight > 0 &&
                    result.rejectWeight / votedWeight > 1 - threshold) {
                    return (result.rejectWeight / (votedWeight + remainingWeight) > 1 - threshold);
                }
                return false;
            }
            case "majority":
            case "weighted":
                return (result.approveWeight > totalWeight / 2 ||
                    result.rejectWeight > totalWeight / 2);
            case "quorum":
                if (!result.quorumMet)
                    return false;
                return (result.approveWeight > totalWeight / 2 ||
                    result.rejectWeight > totalWeight / 2);
            default:
                return false;
        }
    }
    // ── Weight helpers ───────────────────────────────────────────────────────
    getAgentWeight(proposal, agent) {
        if (proposal.weights) {
            const w = proposal.weights.find((w) => w.agent === agent);
            if (w)
                return w.weight;
        }
        return 1;
    }
    getTotalWeight(proposal) {
        let total = 0;
        for (const p of proposal.participants) {
            total += this.getAgentWeight(proposal, p);
        }
        return total;
    }
    // ── Lifecycle ────────────────────────────────────────────────────────────
    resolveProposal(proposal, result) {
        proposal.status =
            result.decision === "approved"
                ? "approved"
                : result.decision === "rejected"
                    ? "rejected"
                    : "rejected"; // no_consensus maps to rejected (not expired — that's for timeouts)
        proposal.result = result;
        this.clearExpiryTimer(proposal.id);
        this.emit("proposal:resolved", proposal, result);
        this.evictOldProposals();
    }
    expireProposal(proposal) {
        if (proposal.status !== "pending")
            return;
        const result = this.calculateResult(proposal);
        proposal.status = "expired";
        proposal.result = result;
        this.clearExpiryTimer(proposal.id);
        this.emit("proposal:expired", proposal);
        this.evictOldProposals();
    }
    scheduleExpiry(proposal) {
        const timeoutMs = proposal.expiresAt - Date.now();
        if (timeoutMs <= 0) {
            this.expireProposal(proposal);
            return;
        }
        const timer = setTimeout(() => this.expireProposal(proposal), timeoutMs);
        timer.unref();
        this.expiryTimers.set(proposal.id, timer);
    }
    clearExpiryTimer(proposalId) {
        const timer = this.expiryTimers.get(proposalId);
        if (timer) {
            clearTimeout(timer);
            this.expiryTimers.delete(proposalId);
        }
    }
    /** Remove oldest resolved/expired/cancelled proposals when over the limit. */
    evictOldProposals() {
        const max = this.config.maxRetainedProposals;
        if (max <= 0)
            return; // unlimited
        const terminal = [];
        for (const p of this.proposals.values()) {
            if (p.status !== "pending")
                terminal.push(p);
        }
        if (terminal.length <= max)
            return;
        // Sort oldest-first by createdAt, evict the excess
        terminal.sort((a, b) => a.createdAt - b.createdAt);
        const toEvict = terminal.length - max;
        for (let i = 0; i < toEvict; i++) {
            this.proposals.delete(terminal[i].id);
        }
    }
    cleanup() {
        for (const timer of this.expiryTimers.values())
            clearTimeout(timer);
        this.expiryTimers.clear();
    }
    // ── Stats ────────────────────────────────────────────────────────────────
    getStats() {
        let pending = 0, approved = 0, rejected = 0, expired = 0, cancelled = 0;
        let totalParticipation = 0;
        let resolvedCount = 0;
        for (const p of this.proposals.values()) {
            if (p.status === "pending")
                pending++;
            else if (p.status === "approved")
                approved++;
            else if (p.status === "rejected")
                rejected++;
            else if (p.status === "expired")
                expired++;
            else if (p.status === "cancelled")
                cancelled++;
            if (p.result) {
                totalParticipation += p.result.participation;
                resolvedCount++;
            }
        }
        return {
            total: this.proposals.size,
            pending,
            approved,
            rejected,
            expired,
            cancelled,
            avgParticipation: resolvedCount > 0 ? totalParticipation / resolvedCount : 0,
        };
    }
}
// ── Factory ──────────────────────────────────────────────────────────────────
export function createConsensusEngine(config) {
    return new ConsensusEngine(config);
}
