// @agent-relay/sdk/dist/http.js
/**
 * Lightweight HTTP client for Relaycast API — safe for bundlers (Vite, Rollup, Webpack).
 * No native dependencies, no broker binary, no process spawning.
 *
 * Usage:
 *   import { RelayCast, createWorkspace, registerAgent } from '@agent-relay/sdk/http';
 */
// Re-export the HTTP-safe parts of @relaycast/sdk
export { RelayCast, RelayError } from '@relaycast/sdk';
const DEFAULT_BASE_URL = 'https://api.relaycast.dev';
/**
 * Create a new Relaycast workspace using plain fetch.
 * This function has zero dependencies beyond the Fetch API.
 *
 * @param name - Human-readable workspace name
 * @param baseUrl - API base URL (defaults to https://api.relaycast.dev)
 * @returns The workspace ID and API key
 */
export async function createWorkspace(name, baseUrl = DEFAULT_BASE_URL) {
    const url = `${baseUrl}/v1/workspaces`;
    const response = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ name }),
    });
    if (!response.ok) {
        const body = await response.text().catch(() => '');
        throw new Error(`Failed to create workspace (HTTP ${response.status}): ${body}`);
    }
    const body = (await response.json());
    // The API wraps responses in { ok, data } — unwrap transparently
    return 'data' in body && body.data ? body.data : body;
}
/**
 * Register an agent in a workspace using plain fetch.
 * This function has zero dependencies beyond the Fetch API.
 *
 * @param apiKey - Workspace API key
 * @param name - Agent name to register
 * @param baseUrl - API base URL (defaults to https://api.relaycast.dev)
 * @returns The registered agent name and token
 */
export async function registerAgent(apiKey, name, baseUrl = DEFAULT_BASE_URL) {
    const url = `${baseUrl}/v1/agents`;
    const response = await fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${apiKey}`,
        },
        body: JSON.stringify({ name }),
    });
    if (!response.ok) {
        const errBody = await response.text().catch(() => '');
        throw new Error(`Failed to register agent (HTTP ${response.status}): ${errBody}`);
    }
    const body = (await response.json());
    return 'data' in body && body.data ? body.data : body;
}
