// @agent-relay/sdk/dist/communicate/adapters/claude-sdk.js
import { Relay } from '../core.js';
import { formatRelayMessage } from '../types.js';
function appendHook(hooks, event, callback) {
    const matchers = [...(hooks?.[event] ?? []), { hooks: [callback] }];
    return {
        ...hooks,
        [event]: matchers,
    };
}
async function drainInbox(relay) {
    const messages = await relay.inbox();
    if (messages.length === 0) {
        return undefined;
    }
    return `New messages from other agents:\n${messages.map((message) => formatRelayMessage(message)).join('\n')}`;
}
/**
 * Attach relay communication to Claude Agent SDK options via MCP server and hooks.
 * @param _name - Agent name for relay registration.
 * @param options - Claude SDK query options to augment.
 * @param relay - Optional pre-configured Relay instance.
 * @returns Augmented options with relaycast MCP server and inbox-draining hooks.
 */
export function onRelay(_name, options, relay = new Relay(_name)) {
    const mcpServers = {
        ...(options.mcpServers ?? {}),
        relaycast: {
            command: 'agent-relay',
            args: ['mcp'],
        },
    };
    let hooks = options.hooks;
    hooks = appendHook(hooks, 'PostToolUse', async () => {
        const systemMessage = await drainInbox(relay);
        return systemMessage ? { systemMessage } : {};
    });
    hooks = appendHook(hooks, 'Stop', async () => {
        const systemMessage = await drainInbox(relay);
        return systemMessage ? { continue: true, systemMessage } : {};
    });
    return {
        ...options,
        mcpServers,
        hooks,
    };
}
