// @agent-relay/sdk/dist/communicate/adapters/crewai.js
import { Relay } from '../core.js';
import { formatRelayInbox, formatRelayMessage } from '../types.js';
function createRelayTools(relay) {
    return [
        {
            tool_name: 'relay_send',
            description: 'Send a direct message to another relay agent.',
            async execute(args) {
                await relay.send(args.to, args.text);
                return `Sent relay message to ${args.to}.`;
            },
        },
        {
            tool_name: 'relay_inbox',
            description: 'Drain and inspect newly received relay messages.',
            async execute() {
                return formatRelayInbox(await relay.inbox());
            },
        },
        {
            tool_name: 'relay_post',
            description: 'Post a message to a relay channel.',
            async execute(args) {
                await relay.post(args.channel, args.text);
                return `Posted relay message to #${args.channel}.`;
            },
        },
        {
            tool_name: 'relay_agents',
            description: 'List currently online relay agents.',
            async execute() {
                return (await relay.agents()).join('\n');
            },
        },
    ];
}
/**
 * Attach relay communication tools and message routing to a CrewAI Agent.
 *
 * Adds relay tools to the agent's `tools` array and installs a `step_callback`
 * that receives incoming relay messages.
 *
 * @param agent - A CrewAI Agent instance (or duck-typed equivalent).
 * @param relay - Optional pre-configured Relay instance.
 * @returns Object with an `unsubscribe` function to stop message routing.
 */
export function onRelay(agent, relay) {
    const relayInstance = relay ?? new Relay(agent.role);
    const relayTools = createRelayTools(relayInstance);
    // Append relay tools to agent's tools array
    for (const tool of relayTools) {
        agent.tools.push(tool);
    }
    // Install message routing via step_callback
    const originalCallback = agent.step_callback;
    const unsubscribe = relayInstance.onMessage(async (message) => {
        const formatted = formatRelayMessage(message);
        // Route relay messages through the original callback if one exists
        if (originalCallback) {
            originalCallback({ relay_message: formatted });
        }
    });
    // Compose original callback with relay step forwarding
    if (originalCallback) {
        agent.step_callback = (step) => {
            // Forward non-relay steps to original callback
            if (!step?.relay_message) {
                originalCallback(step);
            }
        };
    }
    return { unsubscribe };
}
/**
 * Attach relay communication tools to all agents in a CrewAI Crew.
 *
 * @param crew - A CrewAI Crew instance (or duck-typed equivalent).
 * @param relay - Optional pre-configured Relay instance.
 * @returns Object with an `unsubscribe` function to stop all message routing.
 */
export function onCrewRelay(crew, relay) {
    const unsubscribers = [];
    for (const agent of crew.agents) {
        const result = onRelay(agent, relay);
        unsubscribers.push(result.unsubscribe);
    }
    return {
        unsubscribe() {
            for (const unsub of unsubscribers) {
                unsub();
            }
        },
    };
}
