// @agent-relay/sdk/dist/communicate/adapters/openai-agents.js
import { Relay } from '../core.js';
import { formatRelayInbox, formatRelayMessage } from '../types.js';
function schema(props, required) {
    return { type: 'object', properties: props, required, additionalProperties: false };
}
function createRelayTools(relay) {
    const defs = [
        ['relay_send', 'Send a direct message to another relay agent.',
            schema({ to: { type: 'string' }, text: { type: 'string' } }, ['to', 'text']),
            async (p) => { await relay.send(p.to, p.text); return `Sent relay message to ${p.to}.`; }],
        ['relay_inbox', 'Drain and inspect newly received relay messages.',
            schema({}, []),
            async () => formatRelayInbox(await relay.inbox())],
        ['relay_post', 'Post a message to a relay channel.',
            schema({ channel: { type: 'string' }, text: { type: 'string' } }, ['channel', 'text']),
            async (p) => { await relay.post(p.channel, p.text); return `Posted relay message to #${p.channel}.`; }],
        ['relay_agents', 'List currently online relay agents.',
            schema({}, []),
            async () => (await relay.agents()).join('\n')],
    ];
    return defs.map(([name, description, parameters, run]) => ({
        type: 'function',
        name,
        description,
        parameters,
        strict: false,
        needsApproval: false,
        isEnabled: true,
        async invoke(_runContext, input) {
            const params = input ? JSON.parse(input) : {};
            return run(params);
        },
    }));
}
/**
 * Attach relay communication tools and message routing to an OpenAI Agent.
 * @param agent - OpenAI Agent instance to augment with relay tools.
 * @param relay - Optional pre-configured Relay instance.
 * @returns Object with the augmented agent and a cleanup function.
 */
export function onRelay(agent, relay = new Relay(agent.name)) {
    const relayTools = createRelayTools(relay);
    agent.tools = [...agent.tools, ...relayTools];
    const pendingMessages = [];
    const originalInstructions = agent.instructions;
    const unsubscribe = relay.onMessage(async (message) => {
        pendingMessages.push(formatRelayMessage(message));
    });
    agent.instructions = async (...args) => {
        let base;
        if (typeof originalInstructions === 'function') {
            base = await originalInstructions(...args);
        }
        else {
            base = originalInstructions ?? '';
        }
        if (pendingMessages.length > 0) {
            const msgs = pendingMessages.splice(0, pendingMessages.length);
            return `${base}\n\n--- Relay Messages ---\n${msgs.join('\n')}`;
        }
        return base;
    };
    const cleanup = () => {
        unsubscribe();
        agent.instructions = originalInstructions;
        agent.tools = agent.tools.filter((t) => !relayTools.includes(t));
    };
    return { agent, cleanup };
}
