// @agent-relay/sdk/dist/workflows/api-executor.js
/**
 * API Executor — calls LLM provider APIs directly via fetch().
 * Used when agent cli is 'api'. No sandbox, no CLI, no PTY.
 */
function detectProvider(model) {
    if (model.startsWith('claude'))
        return 'anthropic';
    if (model.startsWith('gpt') || model.startsWith('o1') || model.startsWith('o3') || model.startsWith('o4'))
        return 'openai';
    if (model.startsWith('gemini'))
        return 'google';
    return 'anthropic';
}
function getApiKey(provider, envSecrets) {
    const envMap = {
        anthropic: ['ANTHROPIC_API_KEY'],
        openai: ['OPENAI_API_KEY'],
        google: ['GOOGLE_API_KEY', 'GEMINI_API_KEY'],
    };
    for (const key of envMap[provider]) {
        const value = envSecrets?.[key] ?? process.env[key];
        if (value)
            return value;
    }
    throw new Error(`No API key for "${provider}". Set ${envMap[provider].join(' or ')}.`);
}
async function callAnthropic(apiKey, model, task, maxTokens, systemPrompt) {
    const res = await fetch('https://api.anthropic.com/v1/messages', {
        method: 'POST',
        headers: { 'content-type': 'application/json', 'x-api-key': apiKey, 'anthropic-version': '2023-06-01' },
        body: JSON.stringify({
            model, max_tokens: maxTokens,
            ...(systemPrompt ? { system: systemPrompt } : {}),
            messages: [{ role: 'user', content: task }],
        }),
    });
    if (!res.ok)
        throw new Error(`Anthropic API error (${res.status}): ${await res.text()}`);
    const data = await res.json();
    return {
        content: data.content.filter(c => c.type === 'text').map(c => c.text ?? '').join(''),
        model: data.model,
        usage: data.usage ? { inputTokens: data.usage.input_tokens, outputTokens: data.usage.output_tokens } : undefined,
    };
}
async function callOpenAI(apiKey, model, task, maxTokens, systemPrompt) {
    const messages = [];
    if (systemPrompt)
        messages.push({ role: 'system', content: systemPrompt });
    messages.push({ role: 'user', content: task });
    const res = await fetch('https://api.openai.com/v1/chat/completions', {
        method: 'POST',
        headers: { 'content-type': 'application/json', 'authorization': `Bearer ${apiKey}` },
        body: JSON.stringify({ model, max_tokens: maxTokens, messages }),
    });
    if (!res.ok)
        throw new Error(`OpenAI API error (${res.status}): ${await res.text()}`);
    const data = await res.json();
    return {
        content: data.choices[0]?.message?.content ?? '',
        model: data.model,
        usage: data.usage ? { inputTokens: data.usage.prompt_tokens, outputTokens: data.usage.completion_tokens } : undefined,
    };
}
async function callGoogle(apiKey, model, task, maxTokens, systemPrompt) {
    const res = await fetch(`https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent`, {
        method: 'POST',
        headers: { 'content-type': 'application/json', 'x-goog-api-key': apiKey },
        body: JSON.stringify({
            ...(systemPrompt ? { systemInstruction: { parts: [{ text: systemPrompt }] } } : {}),
            contents: [{ parts: [{ text: task }] }],
            generationConfig: { maxOutputTokens: maxTokens },
        }),
    });
    if (!res.ok)
        throw new Error(`Google API error (${res.status}): ${await res.text()}`);
    const data = await res.json();
    return {
        content: data.candidates[0]?.content?.parts?.map(p => p.text).join('') ?? '',
        model,
        usage: data.usageMetadata ? { inputTokens: data.usageMetadata.promptTokenCount, outputTokens: data.usageMetadata.candidatesTokenCount } : undefined,
    };
}
const PROVIDER_CALLERS = { anthropic: callAnthropic, openai: callOpenAI, google: callGoogle };
export async function executeApiStep(model, task, options = {}) {
    const resolvedModel = model || options.defaultModel || 'claude-sonnet-4-20250514';
    const maxTokens = options.defaultMaxTokens ?? 4096;
    const provider = detectProvider(resolvedModel);
    const apiKey = getApiKey(provider, options.envSecrets);
    const response = await PROVIDER_CALLERS[provider](apiKey, resolvedModel, task, maxTokens, options.skills);
    return response.content;
}
export { detectProvider, getApiKey };
