// @agent-relay/sdk/dist/workflows/barrier.js
/**
 * Barrier Manager — synchronization barriers with all/any/majority semantics.
 *
 * Barriers gate downstream workflow steps until a set of upstream agents
 * or steps have resolved. Supports three resolution modes:
 *
 * - **all**      — every agent in `waitFor` must resolve (default)
 * - **any**      — at least one agent resolves
 * - **majority** — more than half of `waitFor` must resolve
 */
import { randomBytes } from 'node:crypto';
import { EventEmitter } from 'node:events';
// ── Manager ─────────────────────────────────────────────────────────────────
export class BarrierManager extends EventEmitter {
    db;
    /** In-memory mode tracking (not persisted — set once at creation). */
    modes = new Map();
    timeoutTimers = new Map();
    constructor(db) {
        super();
        this.db = db;
    }
    // ── Create ──────────────────────────────────────────────────────────────
    /**
     * Create a barrier for a workflow run.
     */
    async createBarrier(runId, definition) {
        const id = `bar_${Date.now()}_${randomBytes(4).toString('hex')}`;
        const now = new Date().toISOString();
        const mode = definition.mode ?? 'all';
        const { rows } = await this.db.query(`INSERT INTO workflow_barriers (id, run_id, barrier_name, wait_for, resolved, is_satisfied, timeout_ms, created_at, updated_at)
       VALUES ($1, $2, $3, $4, '[]'::jsonb, FALSE, $5, $6, $6)
       RETURNING *`, [
            id,
            runId,
            definition.name,
            JSON.stringify(definition.waitFor),
            definition.timeoutMs ?? null,
            now,
        ]);
        const barrier = rows[0];
        const key = `${runId}:${definition.name}`;
        this.modes.set(key, mode);
        if (definition.timeoutMs) {
            this.scheduleTimeout(barrier, definition.timeoutMs);
        }
        this.emit('barrier:created', barrier);
        return barrier;
    }
    /**
     * Bulk-create barriers from a list of definitions (e.g. from coordination config).
     */
    async createBarriers(runId, definitions) {
        const results = [];
        for (const def of definitions) {
            results.push(await this.createBarrier(runId, def));
        }
        return results;
    }
    // ── Resolve ─────────────────────────────────────────────────────────────
    /**
     * Mark an agent/step as resolved for a barrier. Returns whether the
     * barrier is now fully satisfied.
     */
    async resolve(runId, barrierName, agent) {
        const now = new Date().toISOString();
        // Atomic: append agent to resolved array if not already present.
        const { rows } = await this.db.query(`UPDATE workflow_barriers
       SET resolved = CASE
             WHEN resolved @> $3::jsonb THEN resolved
             ELSE resolved || $3::jsonb
           END,
           updated_at = $4
       WHERE run_id = $1 AND barrier_name = $2 AND is_satisfied = FALSE
       RETURNING *`, [runId, barrierName, JSON.stringify(agent), now]);
        if (rows.length === 0) {
            // Barrier may already be satisfied or not exist.
            const existing = await this.getBarrier(runId, barrierName);
            if (!existing)
                throw new Error(`Barrier ${barrierName} not found for run ${runId}`);
            return { satisfied: existing.isSatisfied, barrier: existing };
        }
        const barrier = rows[0];
        this.emit('barrier:resolved', barrierName, agent);
        const key = `${runId}:${barrierName}`;
        const mode = this.modes.get(key) ?? 'all';
        if (this.checkSatisfied(barrier, mode)) {
            return this.markSatisfied(barrier);
        }
        return { satisfied: false, barrier };
    }
    // ── Queries ─────────────────────────────────────────────────────────────
    async getBarrier(runId, barrierName) {
        const { rows } = await this.db.query(`SELECT * FROM workflow_barriers WHERE run_id = $1 AND barrier_name = $2`, [runId, barrierName]);
        return rows[0] ?? null;
    }
    async getBarriers(runId) {
        const { rows } = await this.db.query(`SELECT * FROM workflow_barriers WHERE run_id = $1 ORDER BY created_at ASC`, [runId]);
        return rows;
    }
    async getUnsatisfiedBarriers(runId) {
        const { rows } = await this.db.query(`SELECT * FROM workflow_barriers WHERE run_id = $1 AND is_satisfied = FALSE ORDER BY created_at ASC`, [runId]);
        return rows;
    }
    /**
     * Check if a named barrier is satisfied (useful for gating downstream work).
     */
    async isSatisfied(runId, barrierName) {
        const barrier = await this.getBarrier(runId, barrierName);
        return barrier?.isSatisfied ?? false;
    }
    // ── Cleanup ─────────────────────────────────────────────────────────────
    cleanup() {
        for (const timer of this.timeoutTimers.values())
            clearTimeout(timer);
        this.timeoutTimers.clear();
        this.modes.clear();
    }
    // ── Private ─────────────────────────────────────────────────────────────
    checkSatisfied(barrier, mode) {
        const waitFor = Array.isArray(barrier.waitFor) ? barrier.waitFor : [];
        const resolved = Array.isArray(barrier.resolved) ? barrier.resolved : [];
        switch (mode) {
            case 'all':
                return waitFor.every((w) => resolved.includes(w));
            case 'any':
                return resolved.length > 0;
            case 'majority':
                return resolved.length > waitFor.length / 2;
        }
    }
    async markSatisfied(barrier) {
        const now = new Date().toISOString();
        const { rows } = await this.db.query(`UPDATE workflow_barriers SET is_satisfied = TRUE, updated_at = $2
       WHERE id = $1
       RETURNING *`, [barrier.id, now]);
        const updated = rows[0];
        const key = `${barrier.runId}:${barrier.barrierName}`;
        this.clearTimeout(key);
        this.emit('barrier:satisfied', updated);
        return { satisfied: true, barrier: updated };
    }
    scheduleTimeout(barrier, timeoutMs) {
        const key = `${barrier.runId}:${barrier.barrierName}`;
        const timer = setTimeout(async () => {
            const current = await this.getBarrier(barrier.runId, barrier.barrierName);
            if (current && !current.isSatisfied) {
                this.emit('barrier:timeout', current);
            }
        }, timeoutMs);
        timer.unref();
        this.timeoutTimers.set(key, timer);
    }
    clearTimeout(key) {
        const timer = this.timeoutTimers.get(key);
        if (timer) {
            globalThis.clearTimeout(timer);
            this.timeoutTimers.delete(key);
        }
    }
}
