// @agent-relay/sdk/dist/workflows/cloud-runner.js
export async function runInCloud(config, options) {
    const { cloudApiUrl, cloudApiToken, envSecrets, pollIntervalMs = 3000, timeoutMs = 1800000 } = options;
    const baseUrl = cloudApiUrl.replace(/\/$/, '');
    const { stringify: stringifyYaml } = await import('yaml');
    const yamlStr = stringifyYaml(config);
    const submitRes = await fetch(`${baseUrl}/api/v1/workflows/run`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${cloudApiToken}` },
        body: JSON.stringify({ workflow: yamlStr, fileType: 'yaml', ...(envSecrets ? { envSecrets } : {}) }),
    });
    if (!submitRes.ok)
        throw new Error(`Cloud submit failed (${submitRes.status}): ${await submitRes.text()}`);
    const { runId } = (await submitRes.json());
    const deadline = Date.now() + timeoutMs;
    let lastStatus = 'pending';
    while (Date.now() < deadline) {
        await new Promise(r => setTimeout(r, pollIntervalMs));
        const statusRes = await fetch(`${baseUrl}/api/v1/workflows/runs/${runId}`, {
            headers: { 'Authorization': `Bearer ${cloudApiToken}` },
        });
        if (!statusRes.ok)
            continue;
        const data = (await statusRes.json());
        if (data.status !== lastStatus) {
            lastStatus = data.status;
            options.onStatusChange?.(lastStatus, runId);
        }
        if (data.status === 'completed' || data.status === 'failed') {
            return {
                id: runId, workspaceId: '', workflowName: config.name ?? 'cloud-workflow',
                pattern: config.swarm?.pattern ?? 'dag', status: data.status, config,
                startedAt: data.createdAt ?? new Date().toISOString(),
                completedAt: data.updatedAt ?? new Date().toISOString(),
                error: data.error, createdAt: data.createdAt ?? new Date().toISOString(),
                updatedAt: data.updatedAt ?? new Date().toISOString(),
            };
        }
    }
    throw new Error(`Cloud workflow timed out after ${timeoutMs}ms (runId: ${runId})`);
}
