// @agent-relay/sdk/dist/workflows/memory-db.js
/**
 * In-memory implementation of WorkflowDb for local workflow runs.
 * No persistence — state lives only for the duration of the process.
 */
export class InMemoryWorkflowDb {
    runs = new Map();
    steps = new Map();
    async insertRun(run) {
        this.runs.set(run.id, { ...run });
    }
    async updateRun(id, patch) {
        const existing = this.runs.get(id);
        if (!existing)
            return;
        this.runs.set(id, { ...existing, ...patch, updatedAt: new Date().toISOString() });
    }
    async getRun(id) {
        return this.runs.get(id) ?? null;
    }
    async insertStep(step) {
        this.steps.set(step.id, { ...step });
    }
    async updateStep(id, patch) {
        const existing = this.steps.get(id);
        if (!existing)
            return;
        this.steps.set(id, { ...existing, ...patch, updatedAt: new Date().toISOString() });
    }
    async getStepsByRunId(runId) {
        return Array.from(this.steps.values()).filter((s) => s.runId === runId);
    }
}
