// @relaycast/sdk/dist/agent.js
import { WsClient, withInternalWsOrigin } from './ws.js';
function stripHash(channel) {
    return channel.startsWith('#') ? channel.slice(1) : channel;
}
function normalizeAutoHeartbeatMs(value) {
    if (value === false)
        return false;
    if (typeof value === 'number' && Number.isFinite(value) && value > 0) {
        return Math.floor(value);
    }
    return 30_000;
}
function generateIdempotencyKey() {
    const randomUuid = globalThis.crypto?.randomUUID?.();
    if (typeof randomUuid === 'string' && randomUuid.length > 0) {
        return randomUuid;
    }
    return `${Date.now()}-${Math.random().toString(16).slice(2)}-${Math.random().toString(16).slice(2)}`;
}
function idempotencyHeaders(opts) {
    const key = opts?.idempotencyKey?.trim() || generateIdempotencyKey();
    return {
        headers: {
            'Idempotency-Key': key,
            'X-Idempotency-Key': key,
        },
    };
}
export class AgentClient {
    client;
    ws = null;
    autoHeartbeatMs;
    autoHeartbeatTimer = null;
    wsOptions;
    constructor(client, options = {}) {
        this.client = client;
        this.autoHeartbeatMs = normalizeAutoHeartbeatMs(options.autoHeartbeatMs);
        this.wsOptions = options.ws ?? {};
    }
    /**
     * Presence lifecycle ownership:
     * - worker identities should call `markOnline`/`heartbeat`/`markOffline`.
     * - broker identities should signal presence only when they directly own worker lifecycle.
     * - reader identities should not publish presence.
     */
    presence = {
        markOnline: async () => {
            await this.client.post('/v1/agents/heartbeat', {});
        },
        heartbeat: async () => {
            await this.client.post('/v1/agents/heartbeat', {});
        },
        markOffline: async () => {
            this.stopAutoHeartbeat();
            await this.client.post('/v1/agents/disconnect', {});
        },
    };
    startAutoHeartbeat() {
        this.stopAutoHeartbeat();
        if (this.autoHeartbeatMs === false)
            return;
        this.autoHeartbeatTimer = setInterval(() => {
            void this.presence.heartbeat().catch(() => { });
        }, this.autoHeartbeatMs);
    }
    stopAutoHeartbeat() {
        if (this.autoHeartbeatTimer) {
            clearInterval(this.autoHeartbeatTimer);
            this.autoHeartbeatTimer = null;
        }
    }
    // === WebSocket ===
    connect() {
        if (this.ws)
            return;
        this.ws = new WsClient(withInternalWsOrigin({
            token: this.client.apiKey,
            baseUrl: this.client.baseUrl,
            ...this.wsOptions,
        }, {
            surface: this.client.originSurface,
            client: this.client.originClient,
            version: this.client.originVersion,
        }));
        this.ws.on('open', () => {
            void this.presence.markOnline().catch(() => { });
            this.startAutoHeartbeat();
        });
        this.ws.on('close', () => {
            this.stopAutoHeartbeat();
        });
        this.ws.on('permanently_disconnected', () => {
            this.stopAutoHeartbeat();
        });
        this.ws.connect();
    }
    /** Send a REST heartbeat to keep this agent online in PresenceDO without a WebSocket. */
    async heartbeat() {
        await this.presence.heartbeat();
    }
    async disconnect() {
        this.stopAutoHeartbeat();
        if (this.ws) {
            // Notify the server before closing the WebSocket so presence
            // updates immediately (works around local DO hibernation issues).
            await this.presence.markOffline().catch(() => { });
            this.ws.disconnect();
            this.ws = null;
        }
    }
    subscribe(channels) {
        this.ws?.subscribe(channels);
    }
    unsubscribe(channels) {
        this.ws?.unsubscribe(channels);
    }
    onEvent(eventType, handler) {
        if (!this.ws) {
            throw new Error('WebSocket not connected. Call connect() first.');
        }
        return this.ws.on(eventType, handler);
    }
    on = {
        messageCreated: (handler) => this.onEvent('message.created', handler),
        messageUpdated: (handler) => this.onEvent('message.updated', handler),
        threadReply: (handler) => this.onEvent('thread.reply', handler),
        messageRead: (handler) => this.onEvent('message.read', handler),
        reactionAdded: (handler) => this.onEvent('reaction.added', handler),
        reactionRemoved: (handler) => this.onEvent('reaction.removed', handler),
        dmReceived: (handler) => this.onEvent('dm.received', handler),
        groupDmReceived: (handler) => this.onEvent('group_dm.received', handler),
        agentOnline: (handler) => this.onEvent('agent.online', handler),
        agentOffline: (handler) => this.onEvent('agent.offline', handler),
        channelCreated: (handler) => this.onEvent('channel.created', handler),
        channelUpdated: (handler) => this.onEvent('channel.updated', handler),
        channelArchived: (handler) => this.onEvent('channel.archived', handler),
        memberJoined: (handler) => this.onEvent('member.joined', handler),
        memberLeft: (handler) => this.onEvent('member.left', handler),
        channelMuted: (handler) => this.onEvent('member.channel_muted', handler),
        channelUnmuted: (handler) => this.onEvent('member.channel_unmuted', handler),
        fileUploaded: (handler) => this.onEvent('file.uploaded', handler),
        webhookReceived: (handler) => this.onEvent('webhook.received', handler),
        commandInvoked: (handler) => this.onEvent('command.invoked', handler),
        // Lifecycle
        connected: (handler) => this.onEvent('open', handler),
        disconnected: (handler) => this.onEvent('close', handler),
        error: (handler) => this.onEvent('error', handler),
        reconnecting: (handler) => {
            if (!this.ws) {
                throw new Error('WebSocket not connected. Call connect() first.');
            }
            return this.ws.on('reconnecting', (e) => handler(e.attempt));
        },
        permanentlyDisconnected: (handler) => {
            if (!this.ws) {
                throw new Error('WebSocket not connected. Call connect() first.');
            }
            return this.ws.on('permanently_disconnected', (e) => handler(e.attempt));
        },
        // Wildcard
        any: (handler) => {
            if (!this.ws) {
                throw new Error('WebSocket not connected. Call connect() first.');
            }
            return this.ws.on('*', handler);
        },
    };
    // === Messages ===
    async send(channel, text, opts) {
        const name = stripHash(channel);
        const body = {
            text,
            ...(opts?.attachments ? { attachments: opts.attachments } : {}),
            ...(opts?.blocks ? { blocks: opts.blocks } : {}),
            mode: opts?.mode ?? 'wait',
        };
        return this.client.post(`/v1/channels/${encodeURIComponent(name)}/messages`, body, idempotencyHeaders(opts));
    }
    async messages(channel, opts) {
        const name = stripHash(channel);
        const query = {};
        if (opts?.limit)
            query.limit = String(opts.limit);
        if (opts?.before)
            query.before = opts.before;
        if (opts?.after)
            query.after = opts.after;
        return this.client.get(`/v1/channels/${encodeURIComponent(name)}/messages`, query);
    }
    async message(id) {
        return this.client.get(`/v1/messages/${encodeURIComponent(id)}`);
    }
    async reply(id, text, opts) {
        const body = {
            text,
            ...(opts?.blocks ? { blocks: opts.blocks } : {}),
        };
        return this.client.post(`/v1/messages/${encodeURIComponent(id)}/replies`, body, idempotencyHeaders(opts));
    }
    async thread(id, opts) {
        const query = {};
        if (opts?.limit)
            query.limit = String(opts.limit);
        if (opts?.before)
            query.before = opts.before;
        if (opts?.after)
            query.after = opts.after;
        return this.client.get(`/v1/messages/${encodeURIComponent(id)}/replies`, query);
    }
    // === DMs ===
    async dm(agent, text, opts) {
        const body = {
            to: agent,
            text,
            ...(opts?.attachments ? { attachments: opts.attachments } : {}),
            mode: opts?.mode ?? 'wait',
        };
        return this.client.post('/v1/dm', body, idempotencyHeaders(opts));
    }
    dms = {
        conversations: () => this.client.get('/v1/dm/conversations'),
        messages: (conversationId, opts) => {
            const query = {};
            if (opts?.limit)
                query.limit = String(opts.limit);
            if (opts?.before)
                query.before = opts.before;
            if (opts?.after)
                query.after = opts.after;
            return this.client.get(`/v1/dm/${encodeURIComponent(conversationId)}/messages`, query);
        },
        createGroup: (opts, idempotency) => this.client.post('/v1/dm/group', opts, idempotencyHeaders(idempotency)),
        sendMessage: (conversationId, text, opts) => this.client.post(`/v1/dm/${encodeURIComponent(conversationId)}/messages`, {
            text,
            ...(opts?.attachments ? { attachments: opts.attachments } : {}),
            mode: opts?.mode ?? 'wait',
        }, idempotencyHeaders(opts)),
        addParticipant: (conversationId, agent) => this.client.post(`/v1/dm/${encodeURIComponent(conversationId)}/participants`, { agentName: agent }),
        removeParticipant: (conversationId, agent) => this.client.delete(`/v1/dm/${encodeURIComponent(conversationId)}/participants/${encodeURIComponent(agent)}`),
    };
    // === Channels ===
    channels = {
        create: (data) => this.client.post('/v1/channels', data),
        list: (opts) => {
            const query = {};
            if (opts?.includeArchived)
                query.includeArchived = 'true';
            return this.client.get('/v1/channels', query);
        },
        get: (name) => this.client.get(`/v1/channels/${encodeURIComponent(name)}`),
        join: (name) => this.client.post(`/v1/channels/${encodeURIComponent(name)}/join`),
        leave: async (name) => {
            await this.client.post(`/v1/channels/${encodeURIComponent(name)}/leave`);
        },
        setTopic: (name, topic) => this.client.patch(`/v1/channels/${encodeURIComponent(name)}/topic`, { topic }),
        archive: (name) => this.client.delete(`/v1/channels/${encodeURIComponent(name)}`),
        invite: (channel, agent) => this.client.post(`/v1/channels/${encodeURIComponent(channel)}/invite`, { agentName: agent }),
        members: (name) => this.client.get(`/v1/channels/${encodeURIComponent(name)}/members`),
        update: (name, data) => this.client.patch(`/v1/channels/${encodeURIComponent(name)}`, data),
        mute: async (name) => {
            await this.client.post(`/v1/channels/${encodeURIComponent(name)}/mute`);
        },
        unmute: async (name) => {
            await this.client.post(`/v1/channels/${encodeURIComponent(name)}/unmute`);
        },
    };
    // === Reactions ===
    async react(messageId, emoji) {
        return this.client.post(`/v1/messages/${encodeURIComponent(messageId)}/reactions`, { emoji });
    }
    async unreact(messageId, emoji) {
        return this.client.delete(`/v1/messages/${encodeURIComponent(messageId)}/reactions/${encodeURIComponent(emoji)}`);
    }
    async reactions(messageId) {
        return this.client.get(`/v1/messages/${encodeURIComponent(messageId)}/reactions`);
    }
    // === Search ===
    async search(query, opts) {
        const params = { q: query };
        if (opts?.channel)
            params.channel = opts.channel;
        if (opts?.from)
            params.from = opts.from;
        if (opts?.limit)
            params.limit = String(opts.limit);
        if (opts?.before)
            params.before = opts.before;
        if (opts?.after)
            params.after = opts.after;
        return this.client.get('/v1/search', params);
    }
    // === Inbox ===
    async inbox(options) {
        const params = {};
        if (options?.limit != null)
            params.limit = String(options.limit);
        return this.client.get('/v1/inbox', params);
    }
    // === Read Receipts ===
    async markRead(messageId) {
        return this.client.post(`/v1/messages/${encodeURIComponent(messageId)}/read`);
    }
    async readers(messageId) {
        return this.client.get(`/v1/messages/${encodeURIComponent(messageId)}/readers`);
    }
    async readStatus(channel) {
        const name = stripHash(channel);
        return this.client.get(`/v1/channels/${encodeURIComponent(name)}/read-status`);
    }
    // === Commands ===
    commands = {
        invoke: (command, data) => this.client.post(`/v1/commands/${encodeURIComponent(command)}/invoke`, data),
    };
    // === Files ===
    files = {
        upload: (data) => this.client.post('/v1/files/upload', data),
        complete: (fileId) => this.client.post(`/v1/files/${encodeURIComponent(fileId)}/complete`),
        get: (fileId) => this.client.get(`/v1/files/${encodeURIComponent(fileId)}`),
        delete: (fileId) => this.client.delete(`/v1/files/${encodeURIComponent(fileId)}`),
        list: (opts) => {
            const query = {};
            if (opts?.uploadedBy)
                query.uploadedBy = opts.uploadedBy;
            if (opts?.limit)
                query.limit = String(opts.limit);
            return this.client.get('/v1/files', query);
        },
    };
}
