// @relaycast/sdk/dist/relay.js
import { ApiResponseSchema, CreateWorkspaceResponseSchema, WorkspaceLookupSchema } from '@relaycast/types';
import { AgentClient } from './agent.js';
import { HttpClient } from './client.js';
import { RelayError, relayErrorFromApi } from './errors.js';
import { appendLegacySuffix, emitCompatibilityTelemetry, isNameConflictError, } from './identity.js';
import { SDK_VERSION } from './version.js';
import { SDK_ORIGIN } from './origin.js';
import { camelizeKeys } from './casing.js';
export class RelayCast {
    client;
    identityHint = null;
    workspaceIdHint = null;
    constructor(options) {
        if (!options.apiKey || options.apiKey.trim().length === 0) {
            throw new Error('RelayCast apiKey is required');
        }
        // Preserve hidden internal-origin metadata on options when created via
        // createInternalRelayCast() so downstream requests use the correct origin headers.
        this.client = new HttpClient(options);
    }
    static async createWorkspace(name, baseUrl) {
        const requestBaseUrl = baseUrl ?? 'https://api.relaycast.dev';
        const url = new URL('/v1/workspaces', requestBaseUrl);
        const res = await fetch(url.toString(), {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-SDK-Version': SDK_VERSION,
                'X-Relaycast-Origin-Surface': SDK_ORIGIN.surface,
                'X-Relaycast-Origin-Client': SDK_ORIGIN.client,
                'X-Relaycast-Origin-Version': SDK_ORIGIN.version,
            },
            body: JSON.stringify({ name }),
        });
        let parsed;
        try {
            parsed = await res.json();
        }
        catch (err) {
            throw new RelayError('transport_error', `Failed to parse response as JSON: ${err instanceof Error ? err.message : 'unknown error'}`, { statusCode: res.status, retryable: false, cause: err });
        }
        const envelope = ApiResponseSchema(CreateWorkspaceResponseSchema).safeParse(parsed);
        if (!envelope.success) {
            throw new RelayError('transport_error', 'Response is not a valid Relay API response object', { statusCode: res.status, retryable: false });
        }
        if (!envelope.data.ok) {
            throw relayErrorFromApi(envelope.data.error.code, envelope.data.error.message, res.status);
        }
        return camelizeKeys(envelope.data.data);
    }
    static async lookupWorkspace(name, baseUrl) {
        const requestBaseUrl = baseUrl ?? 'https://api.relaycast.dev';
        const url = new URL(`/v1/workspaces/by-name/${encodeURIComponent(name)}`, requestBaseUrl);
        const res = await fetch(url.toString(), {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                'X-SDK-Version': SDK_VERSION,
                'X-Relaycast-Origin-Surface': SDK_ORIGIN.surface,
                'X-Relaycast-Origin-Client': SDK_ORIGIN.client,
                'X-Relaycast-Origin-Version': SDK_ORIGIN.version,
            },
        });
        let parsed;
        try {
            parsed = await res.json();
        }
        catch (err) {
            throw new RelayError('transport_error', `Failed to parse response as JSON: ${err instanceof Error ? err.message : 'unknown error'}`, { statusCode: res.status, retryable: false, cause: err });
        }
        const envelope = ApiResponseSchema(WorkspaceLookupSchema).safeParse(parsed);
        if (!envelope.success) {
            throw new RelayError('transport_error', 'Response is not a valid Relay API response object', { statusCode: res.status, retryable: false });
        }
        if (!envelope.data.ok) {
            if (res.status === 404) {
                return null;
            }
            throw relayErrorFromApi(envelope.data.error.code, envelope.data.error.message, res.status);
        }
        return camelizeKeys(envelope.data.data);
    }
    static async ensureWorkspace(name, baseUrl) {
        try {
            const created = await RelayCast.createWorkspace(name, baseUrl);
            return { ...created, existed: false, name };
        }
        catch (err) {
            const statusCode = err instanceof RelayError ? err.statusCode : undefined;
            if (statusCode !== 409) {
                throw err;
            }
            const existing = await RelayCast.lookupWorkspace(name, baseUrl);
            if (!existing) {
                throw err;
            }
            return { ...existing, existed: true };
        }
    }
    rememberIdentity(agentId, name) {
        this.identityHint = { agentId, name };
    }
    async resolveWorkspaceId() {
        if (this.workspaceIdHint) {
            return this.workspaceIdHint;
        }
        const workspace = await this.workspace.info();
        this.workspaceIdHint = workspace.id;
        return workspace.id;
    }
    async resolveIdentityInternal() {
        if (!this.identityHint) {
            throw new RelayError('not_found', 'No identity is available. Register or rotate an agent first.', {
                statusCode: 404,
            });
        }
        return {
            agentId: this.identityHint.agentId,
            name: this.identityHint.name,
            workspaceId: await this.resolveWorkspaceId(),
        };
    }
    async registerWithLegacySuffix(data) {
        const maxAttempts = 5;
        let candidateName = data.name;
        for (let attempt = 0; attempt <= maxAttempts; attempt += 1) {
            try {
                return await this.agents.register({ ...data, name: candidateName });
            }
            catch (err) {
                if (!isNameConflictError(err) || attempt === maxAttempts) {
                    throw err;
                }
                candidateName = appendLegacySuffix(data.name);
            }
        }
        throw new RelayError('transport_error', 'Failed to register agent identity after suffix retries');
    }
    registerTypedIdentity(type, data) {
        return this.agents.register({ ...data, type });
    }
    async registerAgent(data) {
        const { strict, ...request } = data;
        if (strict) {
            return this.agents.register(request);
        }
        emitCompatibilityTelemetry('agents.registerAgent.legacy_suffix', {
            requested_name: data.name,
        });
        return this.registerWithLegacySuffix(request);
    }
    async registerOrRotate(data) {
        try {
            return await this.registerAgent({ ...data, strict: true });
        }
        catch (err) {
            if (isNameConflictError(err)) {
                const agent = await this.agents.get(data.name);
                const { token } = await this.agents.rotateToken(agent.name);
                this.rememberIdentity(agent.id, agent.name);
                const createdAt = agent.createdAt ?? agent.lastSeen;
                return {
                    id: agent.id,
                    name: agent.name,
                    token,
                    status: agent.status,
                    createdAt,
                };
            }
            throw err;
        }
    }
    async resolveIdentity() {
        return this.resolveIdentityInternal();
    }
    agent(data) {
        return this.registerTypedIdentity('agent', data);
    }
    human(data) {
        return this.registerTypedIdentity('human', data);
    }
    system(data) {
        return this.registerTypedIdentity('system', data);
    }
    workspace = {
        info: () => this.client.get('/v1/workspace'),
        update: (data) => this.client.patch('/v1/workspace', data),
        delete: () => this.client.delete('/v1/workspace'),
        stream: {
            get: () => this.client.get('/v1/workspace/stream'),
            set: (enabled) => this.client.put('/v1/workspace/stream', { enabled }),
            inherit: () => this.client.put('/v1/workspace/stream', { mode: 'inherit' }),
        },
    };
    systemPrompt = {
        get: () => this.client.get('/v1/workspace/system-prompt'),
        set: (data) => this.client.put('/v1/workspace/system-prompt', data),
    };
    channels = {
        list: (opts) => {
            const query = {};
            if (opts?.includeArchived)
                query.includeArchived = 'true';
            return this.client.get('/v1/channels', query);
        },
        get: (name) => this.client.get(`/v1/channels/${encodeURIComponent(name)}`),
    };
    messages = {
        list: (channel, opts) => {
            const name = channel.startsWith('#') ? channel.slice(1) : channel;
            const query = {};
            if (opts?.limit != null)
                query.limit = String(opts.limit);
            if (opts?.before)
                query.before = opts.before;
            if (opts?.after)
                query.after = opts.after;
            return this.client.get(`/v1/channels/${encodeURIComponent(name)}/messages`, query);
        },
        get: (id) => this.client.get(`/v1/messages/${encodeURIComponent(id)}`),
        thread: (id, opts) => {
            const query = {};
            if (opts?.limit != null)
                query.limit = String(opts.limit);
            if (opts?.before)
                query.before = opts.before;
            if (opts?.after)
                query.after = opts.after;
            return this.client.get(`/v1/messages/${encodeURIComponent(id)}/replies`, query);
        },
        reactions: (id) => this.client.get(`/v1/messages/${encodeURIComponent(id)}/reactions`),
    };
    agents = {
        register: async (data) => {
            const created = await this.client.post('/v1/agents', data);
            this.rememberIdentity(created.id, created.name);
            return created;
        },
        list: (query) => {
            const params = {};
            if (query?.status)
                params.status = query.status;
            return this.client.get('/v1/agents', params);
        },
        get: (name) => this.client.get(`/v1/agents/${encodeURIComponent(name)}`),
        rotateToken: (name) => this.client.post(`/v1/agents/${encodeURIComponent(name)}/rotate-token`, {}),
        update: (name, data) => this.client.patch(`/v1/agents/${encodeURIComponent(name)}`, data),
        delete: (name) => this.client.delete(`/v1/agents/${encodeURIComponent(name)}`),
        presence: () => this.client.get('/v1/agents/presence'),
        registerOrGet: async (data) => {
            emitCompatibilityTelemetry('agents.registerOrGet.deprecated', {
                replacement: 'agents.registerOrRotate',
            });
            return this.registerOrRotate(data);
        },
        registerAgent: (data) => this.registerAgent(data),
        registerOrRotate: (data) => this.registerOrRotate(data),
        resolveIdentity: () => this.resolveIdentity(),
        spawn: (data) => this.client.post('/v1/agents/spawn', data),
        release: (data) => this.client.post('/v1/agents/release', data),
    };
    webhooks = {
        create: (data) => this.client.post('/v1/webhooks', data),
        list: () => this.client.get('/v1/webhooks'),
        delete: (id) => this.client.delete(`/v1/webhooks/${encodeURIComponent(id)}`),
        trigger: (webhookId, data) => this.client.post(`/v1/hooks/${encodeURIComponent(webhookId)}`, data),
    };
    subscriptions = {
        create: (data) => this.client.post('/v1/subscriptions', data),
        list: () => this.client.get('/v1/subscriptions'),
        get: (id) => this.client.get(`/v1/subscriptions/${encodeURIComponent(id)}`),
        delete: (id) => this.client.delete(`/v1/subscriptions/${encodeURIComponent(id)}`),
    };
    commands = {
        register: (data) => this.client.post('/v1/commands', data),
        list: () => this.client.get('/v1/commands'),
        delete: (command) => this.client.delete(`/v1/commands/${encodeURIComponent(command)}`),
    };
    activity = (limit) => {
        const params = {};
        if (limit !== undefined)
            params.limit = String(limit);
        return this.client.get('/v1/activity', params);
    };
    allDmConversations = () => this.client.get('/v1/dm/conversations/all');
    dmMessages = async (conversationId, opts) => {
        const query = {};
        if (opts?.limit !== undefined)
            query.limit = String(opts.limit);
        if (opts?.before)
            query.before = opts.before;
        if (opts?.after)
            query.after = opts.after;
        return this.client.get(`/v1/dm/conversations/${encodeURIComponent(conversationId)}/messages`, query);
    };
    as(agentToken, options) {
        const agentHttpClient = this.client.withApiKey(agentToken);
        return new AgentClient(agentHttpClient, options);
    }
}
